<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Soil eukaryote DNA metabarcoding data from French Guiana — soil_euk • metabaR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Soil eukaryote DNA metabarcoding data from French Guiana — soil_euk"><meta property="og:description" content="Data from a DNA metabarcoding experiment assessing the diversity of soil eukaryotes in two tropical forests in French Guiana."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metabaR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/metabaRF-vignette_karst.html">The karst_euk dataset: sampler effect and technical reproducibility</a>
    </li>
    <li>
      <a href="../articles/metabaRF-vignette.html">Let's metabaR!</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/metabaRfactory/metabaR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Soil eukaryote DNA metabarcoding data from French Guiana</h1>
    <small class="dont-index">Source: <a href="https://github.com/metabaRfactory/metabaR/blob/master/R/soil_euk.R" class="external-link"><code>R/soil_euk.R</code></a></small>
    <div class="hidden name"><code>soil_euk.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Data from a DNA metabarcoding experiment assessing the diversity of soil eukaryotes in two tropical forests in French Guiana.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">soil_euk</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="format">
    <h2>Format</h2>
    <p>An object of class <code>metabarlist</code>; see <code><a href="check_metabarlist.html">check_metabarlist</a></code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Samples were collected at 2 sample sites in contrasting habitats:</p><ul><li><p>The Mana site located in a white sand forest, characterised by highly oligotrophic soils and tree species adaptated to the local harsh conditions.</p></li>
<li><p>The Petit Plateau site is located in pristine rainforest (Nouragues Natural Reserve) characterized by soils rich in clay and organic matter.</p></li>
</ul><p>At each site, sample collection was conducted at 16 sampling points separated from one another by 20 m and arranged in a grid across a 1 ha plot. At each sampling point, two types of environmental material were collected:</p><ul><li><p>soil: a composite sample of 5 soil cores</p></li>
<li><p>litter: 1 m2 of surface leaf litter from the forest floor</p></li>
<li><p>=&gt; A total of 64 DNA extracts (i.e. 16 sampling points x 2 sites x 2 types of environmental matrix, i.e. soil and litter) were thus produced, in addition to four DNA extraction controls (one per site and environmental matrix).</p></li>
</ul><p>For each DNA extract, a short region of the 18S rRNA (primer pair Euka02 in Taberlet et al. 2018) was
amplfied by PCR in quadruplicate, following the protocol described in Zinger et al. (2019).
The resulting PCR products were pooled and sequenced on an Illumina HiSeq platform, using
the paired-end technology.</p>
<p>The total experiment hence resulted in 384 PCR products consisting of:</p><ul><li><p>256 PCR products obtained from biological samples (16 sampling points x 2 sites x 2 environmental matrices x 4 pcr replicates).</p></li>
<li><p>16 PCR products obtained from extraction negative controls (4 extraction negative controls x 4 pcr replicates).</p></li>
<li><p>32 PCR products corresponding to PCR negative controls (8 pcr negative controls x 4 pcr replicates).</p></li>
<li><p>48 sequencing negative controls (i.e. 48 unused tag combinations).</p></li>
<li><p>32 PCR products corresponding to PCR positive controls, i.e. PCR amplifications of a DNA template composed of a mixture of DNA from 16 plant species (8 pcr positive controls at different dilutions x 4 pcr replicates).</p></li>
</ul><p>The retrieved data were then processed using the OBITools (Boyer et al. 2016) and SUMACLUST (Mercier et al. 2013) packages. Briefly, paired-end reads were assembled, assigned to their respective samples/marker and dereplicated. Low-quality sequences (containing Ns, shorter than 50 bp or singletons) were excluded; the remaining sequences were clustered into molecular operational taxonomic units (MOTUs) using SUMACLUST at a sequence similarity threshold of 0.97. The representative sequence of each MOTU (the most abundant one) was assigned a taxonomic clade using a databased built from the EMBL (release 136) with the ecoPCR program (Ficetola et al., 2010). Taxonomic assignments obtained from the SILVAngs pipeline (Quast et al. 2013; using default parameters for the taxonomic identification) are also available at https://github.com/metabaRfactory/metabaR_external_data (lit_euk—ssu* files) .</p>

<p>The data `soil_euk` is a <code>metabarlist</code> containing four tables</p>
<ul><li><p>`reads`: a numeric matrix containing MOTU abundances (expressed as a number of reads) for each PCR (i.e. technical replicates of both biological samples and positive and negative controls)</p></li>
<li><p>`motus`: a dataframe containing MOTU characteristics (e.g. taxonomy, sequence) for each MOTU)</p></li>
<li><p>`pcrs`: a dataframe containing information on each PCR (e.g. control type, PCR well, etc.)</p></li>
<li><p>`samples`: a dataframe containing information on each environmental sample (e.g. habitat type, etc.)</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Boyer, F., Mercier, C., Bonin, A., Le Bras, Y., Taberlet, P., &amp; Coissac, E. (2016). obitools: a unix-inspired software package for DNA metabarcoding. Molecular Ecology Resources, 16(1), 176-182.</p>
<p>Ficetola, G. F., Coissac, E., Zundel, S., Riaz, T., Shehzad, W., Bessière, J., ... &amp; Pompanon, F. (2010). An in silico approach for the evaluation of DNA barcodes. BMC Genomics, 11(1), 434.</p>
<p>Mercier, C., Boyer, F., Bonin, A., &amp; Coissac, E. (2013, November). SUMATRA and SUMACLUST: fast and exact comparison and clustering of sequences. In Programs and Abstracts of the SeqBio 2013 workshop. Abstract (pp. 27-29).</p>
<p>Taberlet, P., Bonin, A., Zinger, L., &amp; Coissac, E. (2018). Environmental DNA: For Biodiversity Research and Monitoring. Oxford University Press.</p>
<p>Zinger, L., Taberlet, P., Schimann, H., Bonin, A., Boyer, F., De Barba, M., ... &amp; Chave, J. (2019). Body size determines soil community assembly in a tropical forest. Molecular Ecology, 28(3), 528-543.</p>
<p>Quast, C., Pruesse, E., Yilmaz, P., Gerken, J., Schweer, T., Yarza, P., ... &amp; Glöckner, F. O. (2012). The SILVA ribosomal RNA gene database project: improved data processing and web-based tools. Nucleic Acids Research, 41(D1), D590-D596.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">soil_euk</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="summary_metabarlist.html">summary_metabarlist</a></span><span class="op">(</span><span class="va">soil_euk</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $dataset_dimension</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         n_row n_col</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reads     384 12647</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> motus   12647    15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> pcrs      384    11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> samples    64     8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $dataset_statistics</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         nb_reads nb_motus avg_reads sd_reads avg_motus sd_motus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> pcrs     3538913    12647  9215.919 10283.45  333.6849  295.440</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> samples  2797294    12382 10926.930 10346.66  489.5117  239.685</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Lucie Zinger, Frédéric Boyer, Clément Lionnet, Anne-Sophie Benoiston.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

