<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Let's metabaR! • metabaR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Let's metabaR!">
<meta property="og:description" content="metabaR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metabaR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/metabaRF-vignette.html">Let's metabaR!</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/metabaRfactory/metabaR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="metabaRF-vignette_files/kePrint-0.0.1/kePrint.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Let’s metabaR!</h1>
                        <h4 class="author">Lucie Zinger, Anne-Sophie Benoiston</h4>
            <address class="author_afil">
      Institut de Biologie de l’ENS (IBENS), Département de biologie, École normale supérieure, CNRS, INSERM, Université PSL, 75005 Paris, France.<br><h4 class="author">Julian Donald</h4>
            <address class="author_afil">
      Evolution et Diversité Biologique (EDB UMR5174), Université Toulouse 3 Paul Sabatier, CNRS, IRD - Toulouse, France.<br><h4 class="author">Clément Lionnet, Frédéric Boyer</h4>
            <address class="author_afil">
      Univ. Grenoble Alpes, CNRS, Univ. Savoie Mont Blanc, LECA, Laboratoire d’Ecologie Alpine, F-38000 Grenoble, France.<br><h4 class="date">2020-08-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/metabaRfactory/metabaR/blob/master/vignettes/metabaRF-vignette.Rmd"><code>vignettes/metabaRF-vignette.Rmd</code></a></small>
      <div class="hidden name"><code>metabaRF-vignette.Rmd</code></div>

    </address>
</address>
</address>
</div>

    
    
<div class="figure">
<img src="../man/figures/metabaR.png" style="width:100.0%">
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><code>metabaR</code> is an R package which supports the importing, handling and post-bioinformatics evaluation and improvement of metabarcoding data quality. It provides a suite of functions to identify and filter common molecular artifacts produced during the experimental workflow, from sampling to sequencing, ideally using experimental controls.</p>
<p>Due to its simple structure, <code>metabaR</code> can easily be used in combination with other R packages commonly used for ecological analysis (<code>vegan</code>, <code>ade4</code>, <code>ape</code>, <code>picante</code>, etc.). In addition, it provides flexible graphical systems based on <code>ggplot2</code> to vizualise data from both an ecological and experimental perspective.</p>
</div>
<div id="dependencies-and-installation" class="section level1">
<h1 class="hasAnchor">
<a href="#dependencies-and-installation" class="anchor"></a>Dependencies and Installation</h1>
<p><code>metabaR</code> relies on basic R functions and data structures so as to maximise fexibility and transposability across other packages. It has a minimal number of dependencies to essential R packages :</p>
<ul>
<li>
<code>ggplot2</code>, <code>cowplot</code>, and <code>igraph</code> for visualisation purposes</li>
<li>
<code>reshape2</code> for data manipulation purposes<br>
</li>
<li>
<code>vegan</code> and <code>ade4</code> for basic data analyses<br>
</li>
<li>
<code>seqinr</code> and <code>biomformat</code> for data import or formatting.</li>
</ul>
<p>To install <code>metabaR</code>, use :</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)
<span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"metabaRactory/metabaR"</span>)</pre></body></html></div>
<p>And then load the package</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">metabaR</span>) <span class="co"># modify the name once we'll all agree on that</span></pre></body></html></div>
</div>
<div id="package-overview" class="section level1">
<h1 class="hasAnchor">
<a href="#package-overview" class="anchor"></a>Package overview</h1>
<div class="figure">
<img src="../man/figures/metabaR_overview.png" style="width:90.0%">
</div>
<div id="data-format-and-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#data-format-and-structure" class="anchor"></a>Data format and structure</h2>
<p>The basic data format used in <code>metabaR</code> is a <code>metabarlist</code>, a list of four tables:</p>
<ul>
<li><p><code>reads</code> a table of class <code>matrix</code> consisting of PCRs as rows, and molecular operational taxonomic units (MOTUs) as columns. The number of reads for each MOTU in each PCR is given in each cell, with 0 corresponding to no reads.</p></li>
<li><p><code>motus</code> a table of class <code>data.frame</code> where MOTUs are listed as rows, and their attributes as columns. <strong>A mandatory field in this table is the field “sequence”, i.e. the DNA sequence representative of the MOTU</strong>. Examples of other attributes that can be included in this table are the MOTU taxonomic information, the taxonomic assignment scores, MOTU sequence GC content, MOTU total abundance in the dataset, etc.</p></li>
<li>
<code>pcrs</code> a table of class <code>data.frame</code> consisting of PCRs as rows, and PCR attributes as columns. This table is particularly important in <code>metabaR</code>, as it contains all the information related to the extraction, PCR and sequencing protocols and design that are necessary to assess and improve the quality of metabarcoding data <span class="citation">(P. Taberlet et al. 2018; Zinger et al. 2019)</span>. This table can also include information relating to the PCR design, such as the well coordinates and plate of each PCR, the tag combinations specific of the PCR, the primers used, etc. <strong>Mandatory fields are</strong>:
<ul>
<li>
<strong><code>sample_id</code></strong>: a vector indicating the biological sample origin of each PCR (e.g. the sample name)</li>
<li>
<strong><code>type</code></strong> : the type of PCR, either a sample or an experimental control amplification. Only two values allowed: <code>"sample"</code> or <code>"control"</code>.<br>
</li>
<li>
<strong><code>control_type</code></strong> : the type of control. Only five values are possible in this field:
<ul>
<li>
<code>NA</code> if <code>type="sample"</code>, i.e. for any PCR obtained from a biological sample.<br>
</li>
<li>
<code>"extraction"</code> for DNA extraction negative controls, i.e. PCR amplification of an extraction where the DNA template was replaced by extraction buffer.<br>
</li>
<li>
<code>"pcr"</code> for PCR negative controls, i.e. pcr amplification where the DNA template was replaced by PCR buffer or sterile water.<br>
</li>
<li>
<code>"sequencing"</code> for sequencing negative controls, i.e. unused tag/library index combinations.<br>
</li>
<li>
<code>"positive"</code> for DNA extraction or PCR positive controls, i.e. pcr amplifications of known biological samples or DNA template (e.g. a mock community).</li>
</ul>
</li>
</ul>
</li>
<li><p><code>samples</code> a table of class <code>data.frame</code> consisting of biological samples as rows, and associated information as columns. Such information includes e.g. geographic coordinates, abiotic parameters, experimental treatment, etc. This table does not include information on the DNA metabarcoding experimental controls, which can only be found in <code>pcrs</code>.</p></li>
</ul>
</div>
<div id="function-types" class="section level2">
<h2 class="hasAnchor">
<a href="#function-types" class="anchor"></a>Function Types</h2>
<p><code>metabaR</code> provides a range of function types:</p>
<ul>
<li>Import and formating functions to import DNA metabarcoding data from common bioinformatic pipelines (e.g. OBITools, data in the biom format) or more generally from any data formatted as described above in 4 tables corresponding to the <code>reads</code>, <code>motus</code>, <code>pcrs</code>, <code>samples</code>.<br>
</li>
<li>Functions for data curation. These are often absent from most bioinformatic pipelines. They aim to detect and enable the tagging of potential molecular artifacts such as contaminants or dysfunctional PCR reactions.<br>
</li>
<li>Functions for visualizing the data under both ecological (e.g. sample types similarity, rarefaction curves) and experimental (e.g. control types, distribution across the PCR plate design) perspectives.<br>
</li>
<li>Functions to manipulate the <code>metabarlist</code> object, such as selection of a subset, data aggregation by PCRs or MOTUs, taxonomy information formatting, etc.</li>
</ul>
</div>
<div id="example-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#example-dataset" class="anchor"></a>Example dataset</h2>
<p>An example dataset is provided in <code>metabaR</code> to show how the package can be used to assess and improve data quality.</p>
<p>The <code>soil_euk</code> dataset is of class <code>metabarlist</code>. The data were obtained from an environmental DNA (eDNA) metabarcoding experiment aiming to assess the diversity of soil eukaryotes in French Guiana in two sites corresponding to two contrasting habitats:<br>
- Mana, a site located in a white sand forest, characterised by highly oligotrophic soils and tree species adapted to the harsh local conditions.<br>
- Petit Plateau, a site located in the pristine rainforest of the Nouragues natural reserve characterised by <em>terra firme</em> soils richer in clay and organic matter.</p>
<p><img src="soil_euk_loc.png" style="width:50.0%"><img src="soil_euk_plots.png" style="width:36.0%"></p>
<p>At each site, both soil and litter samples were collected so as to assess if these two types of material differ in their eukaryotic communities (Figure 2b). The total experiment consists of 384 PCR products, including 256 PCR products obtained from biological samples, the rest corresponding to various experimental controls. The retrieved data were then processed using the OBITools <span class="citation">(F. Boyer et al. 2016)</span> and SUMACLUST <span class="citation">(Mercier et al. 2013)</span> packages, as well as the SILVAngs pipeline <span class="citation">(Quast et al. 2013)</span>.</p>
<p>More information on the <code>soil_euk</code> dataset can be found in its help page:</p>
<div class="sourceCode"><html><body><pre class="r">?<span class="no">soil_euk</span></pre></body></html></div>
<p>The example dataset is loaded in R as follows:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">soil_euk</span>)
<span class="fu"><a href="../reference/summary_metabarlist.html">summary_metabarlist</a></span>(<span class="no">soil_euk</span>)
<span class="co">#&gt; $dataset_dimension</span>
<span class="co">#&gt;         n_row n_col</span>
<span class="co">#&gt; reads     384 12647</span>
<span class="co">#&gt; motus   12647    15</span>
<span class="co">#&gt; pcrs      384    11</span>
<span class="co">#&gt; samples    64     8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $dataset_statistics</span>
<span class="co">#&gt;         nb_reads nb_motus avg_reads sd_reads avg_motus sd_motus</span>
<span class="co">#&gt; pcrs     3538913    12647  9215.919 10283.45  333.6849  295.440</span>
<span class="co">#&gt; samples  2797294    12382 10926.930 10346.66  489.5117  239.685</span></pre></body></html></div>
<p>The <code>summary_metabarlist</code> function displays the dataset dimensions and characteristics. The above shows that the <code>soil_euk</code> dataset is composed of 12647 eukaryote MOTUs from 384 pcrs, corresponding to a total of 64 soil cores plus the different experimental controls (object <code>soil_euk$dataset_dimension</code>). The total number of MOTUs and reads differ between the lines <code>pcrs</code> and <code>samples</code> in the <code>soil_euk$dataset_statistics</code> object because PCRs products (hereafter referred to as <em>pcrs</em>) include experimental positive and negative controls besides biological samples (hereafter referred to as <em>samples</em>).</p>
<p>The <code>soil_euk</code> dataset also contains information relative to MOTUs (15 variables in the <code>motus</code> table), PCRs (11 variables in the <code>pcrs</code> table), and samples (8 variables in the <code>samples</code> table).</p>
<p>Such information can be observed with basic R commands, given that the <code>metabarlist</code> object is a simple R <code>list</code>:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>)
<span class="co">#&gt;  [1] "plate_no"     "plate_col"    "plate_row"    "tag_fwd"      "tag_rev"     </span>
<span class="co">#&gt;  [6] "primer_fwd"   "primer_rev"   "project"      "sample_id"    "type"        </span>
<span class="co">#&gt; [11] "control_type"</span></pre></body></html></div>
<p>In <code>soil_euk$pcrs</code> these columns correspond to:</p>
<ul>
<li>
<code>"plate_no"</code>: the PCR plate number in which each <em>pcr</em> has been conducted.<br>
</li>
<li>
<code>"plate_col"</code> and <code>"plate_row"</code>: the well coordinate (i.e. column and row) corresponding to where each <em>pcr</em> has been conducted.<br>
</li>
<li>
<code>"tag_fwd"</code> and <code>"tag_rev"</code>: the forward and reverse nucleotide tag/indices used to differenciate each <em>pcr</em>.<br>
</li>
<li>
<code>"primer_fwd"</code> and <code>"primer_rev"</code>: the forward and reverse primers used.<br>
</li>
<li>
<code>"project"</code>: the experiment project name.</li>
<li>
<code>"sample_id"</code>, <code>"type"</code>, <code>"control_type"</code>: the mandatory fields for the <code>pcrs</code> table as described above.</li>
</ul>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">soil_euk</span>$<span class="no">samples</span>)
<span class="co">#&gt; [1] "site_id"            "point_id"           "Latitude"          </span>
<span class="co">#&gt; [4] "Longitude"          "Code.Petit.Plateau" "Site"              </span>
<span class="co">#&gt; [7] "Habitat"            "Material"</span></pre></body></html></div>
<p>In <code>soil_euk$samples</code> these columns correspond to:</p>
<ul>
<li>
<code>"site_id"</code> and <code>"point_id"</code>: the identification codes for sites and sampling points respectively.<br>
</li>
<li>
<code>"Latitude"</code> and <code>"Longitude"</code>: the geographic coordinates of the sampling points (decimal degree).<br>
</li>
<li>
<code>"Code.Petit.Plateau"</code>: code specific to the experiment. Useful when several experiments are sequenced in the same library, but to not be considered in this tutorial.<br>
</li>
<li>
<code>"Site"</code> <code>"Habitat"</code>: site name and corresponding habitat of each <em>sample</em>.<br>
</li>
<li>
<code>"Material"</code>: type of substrate material for each <em>sample</em>.</li>
</ul>
</div>
</div>
<div id="tutorial-with-the-soil_euk-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#tutorial-with-the-soil_euk-dataset" class="anchor"></a>Tutorial with the <code>soil_euk</code> dataset</h1>
<p>Below we provide a standard procedure for data analysis and curation with the <code>metabaR</code> package. Note however that not all functions provided in <code>metabaR</code> are indicated below, we encourage the user to explore the package further in order to tailor its use depending on your dataset characteristics.</p>
<div id="data-import" class="section level2">
<h2 class="hasAnchor">
<a href="#data-import" class="anchor"></a>Data import</h2>
<p><code>metabaR</code> provides import tools for different data formats. Let’s consider for example a suite of four classical “.txt” or csv files each corresponding to the future <code>reads</code>, <code>motus</code>, <code>pcrs</code>, and <code>samples</code> objects. These can be imported and formated into a <code>metabarlist</code> with the <code>tabfiles_to_metabarlist</code> function as follows:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">soil_euk</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/tabfiles_to_metabarlist.html">tabfiles_to_metabarlist</a></span>(<span class="kw">file_reads</span> <span class="kw">=</span> <span class="st">"litiere_euk_reads.txt"</span>,
                                    <span class="kw">file_motus</span> <span class="kw">=</span> <span class="st">"litiere_euk_motus.txt"</span>,
                                    <span class="kw">file_pcrs</span> <span class="kw">=</span> <span class="st">"litiere_euk_pcrs.txt"</span>,
                                    <span class="kw">file_samples</span> <span class="kw">=</span> <span class="st">"litiere_euk_samples.txt"</span>)</pre></body></html></div>
<p>The default field separator of input files is tabulation, but this can be modified by the user. The rows x columns of the input files should follow the format of the table <code>reads</code>, <code>motus</code>, <code>pcrs</code>, and <code>samples</code> described above. Note that in this example above, these input files are stored in the current working directory. If you want to use specifically these example files, we refer you to the <code>tabfiles_to_metabarlist</code> help page example.</p>
</div>
<div id="diagnostic-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#diagnostic-plots" class="anchor"></a>Diagnostic Plots</h2>
<p>Before processing the data, it is useful to begin the analysis with an explorative visualisation of the raw data, which can already highlight several potential problems.</p>
<div id="basic-visualisation" class="section level3">
<h3 class="hasAnchor">
<a href="#basic-visualisation" class="anchor"></a>Basic visualisation</h3>
<p>An initial assessment consists in determining how many reads and MOTUs have been obtained across samples and control types, with the expectation that negative controls yield no or much lower read counts and MOTU numbers. To do so, one can either create new independent vectors storing the total number of reads and MOTUs in each <em>pcr</em>, or store these information in the table <code>pcrs</code> of the <code>metabarlist</code> directly, as done below:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Compute the number of reads per pcr</span>
<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">nb_reads</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">soil_euk</span>$<span class="no">reads</span>)

<span class="co"># Compute the number of motus per pcr</span>
<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">nb_motus</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">soil_euk</span>$<span class="no">reads</span><span class="kw">&gt;</span><span class="fl">0</span>)</pre></body></html></div>
<p>And then plot the results using the “control_type” column of the <code>pcrs</code> table</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Load requested package for plotting</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">reshape2</span>)

<span class="co"># Create an input table (named check1) for ggplot of 3 columns: </span>
<span class="co">#  (i) control type </span>
<span class="co">#  (ii) a vector indicated whether it corresponds to nb_reads or nb_motus, </span>
<span class="co">#  (iii) the corresponding values.</span>

<span class="no">check1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"control_type"</span>, <span class="st">"nb_reads"</span>, <span class="st">"nb_motus"</span>)])

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">data</span> <span class="kw">&lt;-</span> <span class="no">check1</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">control_type</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">value</span>, <span class="kw">color</span><span class="kw">=</span><span class="no">control_type</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="kw">alpha</span><span class="kw">=</span><span class="fl">0.2</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"brown"</span>, <span class="st">"red"</span>, <span class="st">"cyan4"</span>,<span class="st">"pink"</span>), <span class="kw">na.value</span> <span class="kw">=</span> <span class="st">"darkgrey"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">variable</span>, <span class="kw">scales</span> <span class="kw">=</span> <span class="st">"free_y"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">axis.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">angle</span><span class="kw">=</span><span class="fl">45</span>, <span class="kw">h</span><span class="kw">=</span><span class="fl">1</span>))</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/diag1_boxplotreadsmotus-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Remember that <em>pcrs</em> obtained from <em>samples</em> are referred to as <code>NA</code> in the <code>control_type</code> vector, shown in grey in the example above. Here, extraction and PCR negative controls yield a few MOTUs of non negligible abundance, which are likely contaminants. Not to worry for now, since this feature is common in DNA metabarcoding datasets <span class="citation">(P. Taberlet et al. 2018; Zinger et al. 2019)</span>, and we will address with those below.</p>
<p>An other basic visualisation consists in determining how the number of MOTUs and reads correlate. This information can help identify the sequencing depth below which a <em>pcr</em> might not be reliable, in particular by comparison with the characteristics of experimental negative controls.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Using the nb_reads and nb_motus defined previously in the soil_euk$pcrs table</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">nb_reads</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">nb_motus</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="no">control_type</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"brown"</span>, <span class="st">"red"</span>, <span class="st">"cyan4"</span>,<span class="st">"pink"</span>), <span class="kw">na.value</span> <span class="kw">=</span> <span class="st">"darkgrey"</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/diag2_readsMOTUs-1.png" width="480" style="display: block; margin: auto;"></p>
<p>In general, we observe a positive correlation between the number of reads and MOTUs per <code>pcr</code>. The strength of this relationship for the different <em>pcrs</em> types (i.e. from <em>samples</em> or controls) can reveal important information about the dataset. A strong correlation suggests that the sequencing effort is not sufficient to cover the <em>pcr</em> diversity. In other words, the number of MOTUs increases linearly with the sequencing depth, which is not a desired property when working with diversity. On the contrary, an absence of correlation suggests that the diversity is well covered by sequencing. The latter is the case in our example above with extraction or pcr negative control <em>pcrs</em> harbouring high number of reads but low numbers of MOTUs: these are most likely contaminants. The read / MOTU number relationship is steepest for sequencing negative controls, which indicates that although tag-jumps do occur at low rates in the <code>soil_euk</code> dataset (low number of reads), it occurs for many MOTUs <span class="citation">(Schnell, Bohmann, and Gilbert 2015)</span>. For <em>pcrs</em> obtained from <em>samples</em>, the number of reads and MOTUs is only slightly correlated. Of note is that some <em>pcrs</em> exhibit features more similar to controls and are likely dysfunctional.</p>
</div>
<div id="visualisation-in-the-pcr-design-context" class="section level3">
<h3 class="hasAnchor">
<a href="#visualisation-in-the-pcr-design-context" class="anchor"></a>Visualisation in the PCR design context</h3>
<p>Another way to vizualise basic descriptive dataset statistics is to represent them in their experimental context. For example, one can plot the number of reads in the PCR plates if the PCR plate numbers and well coordinates have been provided in the <code>pcrs</code> table. Such visualisation can highlight potential issues at the PCR amplification step. For example, low read abundances in <em>pcrs</em> throughout one line or column of the PCR plate suggests that a primer was dysfunctional or that the pipetting of reagents was inconsistent. Let’s see how it appears for the <code>soil_euk</code> data:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/ggpcrplate.html">ggpcrplate</a></span>(<span class="no">soil_euk</span>, <span class="kw">FUN</span> <span class="kw">=</span> <span class="kw">function</span>(<span class="no">m</span>){<span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">m</span>$<span class="no">reads</span>)}, <span class="kw">legend_title</span> <span class="kw">=</span> <span class="st">"# of reads per PCR"</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/ggpcrplate1-1.png" width="672" style="display: block; margin: auto;"></p>
<p><code>ggpcrplate</code> uses a <code>metabarlist</code> and a custom function. In the example above, it computes the number of reads per <em>pcr</em> to display these in their experimental context. This operation is the default when using <code>ggpcrplate</code>, and can be obtained with the bfollowing asic call:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/ggpcrplate.html">ggpcrplate</a></span>(<span class="no">soil_euk</span>)</pre></body></html></div>
<p>Besides the presence of contaminants in extraction and pcr negative controls, one can observe:<br>
- That sequencing negative controls (here corresponding to wells where neither DNA template nor PCR reagents were introduced: these are the unused tag combinations in the <code>soil_euk</code> dataset experimental design), have low or null read numbers. This demonstrates that although “tag-jumps” <span class="citation">(Schnell, Bohmann, and Gilbert 2015; reviewed in Zinger et al. 2019)</span> are present, they remain relatively limited in this experiment.<br>
- That certain lines or columns do exhibit a low number of reads in general, e.g. here in all wells in plate 1, line H. This tendency might denote a pipetting or primer problem, as mentioned above.</p>
<p>If the PCR design uses a combination of two tags as for the <code>soil_euk</code> dataset and that this information is available in the <code>pcrs</code> table, it is also possible to determine if one of the tag introduces systematic biases as follows:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Here the list of all tag/indices used in the experiment </span>
<span class="co"># is available in the column "tag_rev" of the soil_euk$pcrs table</span>
<span class="no">tag.list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">tag_rev</span>))

<span class="fu"><a href="../reference/ggpcrtag.html">ggpcrtag</a></span>(<span class="no">soil_euk</span>,
         <span class="kw">legend_title</span> <span class="kw">=</span> <span class="st">"# of reads per PCR"</span>,
         <span class="kw">FUN</span> <span class="kw">=</span> <span class="kw">function</span>(<span class="no">m</span>) {<span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">m</span>$<span class="no">reads</span>)},
         <span class="kw">taglist</span> <span class="kw">=</span> <span class="no">tag.list</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/ggpcrtag-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The <code>ggpcrtag</code> function works in a similar fashion to <code>ggpcrplate</code>. The plot shows the number of reads in their full PCR design, i.e. with all plates together including their shared tags. The boxplots above and to the right show the distribution of the number of reads obtained for each tag-primer. If the tagging strategy relies on identical tags for both primers, then only the diagonal of this plot will be shown.</p>
</div>
<div id="visualisation-with-tools-borrowed-from-ecology" class="section level3">
<h3 class="hasAnchor">
<a href="#visualisation-with-tools-borrowed-from-ecology" class="anchor"></a>Visualisation with tools borrowed from ecology</h3>
<p>Another useful way to visualise the success of the experiment consists in constructing rarefaction curves, which can indicate whether sequencing effort is sufficient to cover the diversity in MOTUs of each pcr (remember here that sequencing is a sampling process of amplicons/MOTUs within the PCR tube, not of the diversity in the sampling site).</p>
<p>In <code>metabaR</code>, rarefaction curves are constructed with three diversity indices. These are part of the Hill numbers framework <span class="citation">(reviewed in Chao, Chiu, and Jost 2014)</span>, which is based on the concept of effective number species, and is defined as follows:</p>
<p><span class="math inline">\(^{q}D=(\sum_{i=1}^{S}p_i^q)^{1/(1-q)}\)</span></p>
<p>Where <span class="math inline">\(S\)</span> is the total number of species, and <span class="math inline">\(p_i\)</span> the species frequency. The diversity <span class="math inline">\(^{q}D\)</span> corresponds to the diversity of equally-common species, with order <span class="math inline">\(q\)</span> defining the degree of species commonness. In other words, this framework give less weight to rare species when <span class="math inline">\(q\)</span> increases. The interesting feature of the framework is that it unifies mathematically the best known diversity measures in ecology through this unique parameter <span class="math inline">\(q\)</span>:<br>
for <span class="math inline">\(q=0\)</span>, <span class="math inline">\(^{0}D\)</span> is species richness<br>
for <span class="math inline">\(q=1\)</span>, <span class="math inline">\(^{1}D\)</span> approaches the exponential of the Shannon index <span class="math inline">\(H'=\sum_{i=1}^{S}p_i log p_i\)</span><br>
for <span class="math inline">\(q=2\)</span>, <span class="math inline">\(^{2}D\)</span> is the inverted of the Simpson index <span class="math inline">\(\lambda=\sum_{i=1}^{S}p_i^2\)</span><br>
These are the value of <span class="math inline">\(q\)</span> for which <code>metabaR</code> will construct rarefaction curves.</p>
<p>For metabarcoding data, these indices are particularly useful because they allow users to give less weight to rare MOTUs, which often correspond to molecular artifacts and of which inclusion in the data can lead to spurious ecological conclusions <span class="citation">(P. Taberlet et al. 2018; Alberdi and Gilbert 2019; Calderón-Sanou et al. 2020)</span>.</p>
<p><code>metabaR</code> also computes the Good’s coverage index for each pcr:</p>
<p><span class="math inline">\(Coverage = 1-\frac{N_{singletons}}{N}\)</span></p>
<p>Where <span class="math inline">\(N_{singletons}\)</span> is the number of singletons and <span class="math inline">\(N\)</span> the total number of individuals, i.e. the proportion of reads that are not singletons in the <em>pcr</em>. Note that the Good coverage index should be interpreted carefully with DNA metabarcoding data, as it is based on singletons. First, because singletons are potentially spurious signal that can be especially important is species- or DNA-poor samples. This may lead to an overestimation of the coverage. On the other hand, “absolute singletons” (i.e. singletons across the whole sequence dataset) are often filtered out during the bioinformatic process, so the number of singletons observed at this stage of the analysis is likely to be strongly underestimated (and this would artificially lower the coverage estimate).</p>
<p>Since this process can take a while to calculate with large datasets, we here only construct rarefaction curves for a subset of <em>pcrs</em>, in this case by keeping only few <em>samples</em> from the H20 plot of the Petit Plateau. The subsetting of the <code>soil_euk</code> dataset can be done as follows:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Get the samples names from the H20 plot</span>
<span class="no">h20_id</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"H20-[A-B]"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>))

<span class="co"># Subset the data</span>
<span class="no">soil_euk_h20</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/subset_metabarlist.html">subset_metabarlist</a></span>(<span class="no">soil_euk</span>, <span class="kw">table</span> <span class="kw">=</span> <span class="st">"pcrs"</span>, <span class="kw">indices</span> <span class="kw">=</span> <span class="no">h20_id</span>)

<span class="co"># Check results</span>
<span class="fu"><a href="../reference/summary_metabarlist.html">summary_metabarlist</a></span>(<span class="no">soil_euk_h20</span>)
<span class="co">#&gt; $dataset_dimension</span>
<span class="co">#&gt;         n_row n_col</span>
<span class="co">#&gt; reads      16  3182</span>
<span class="co">#&gt; motus    3182    15</span>
<span class="co">#&gt; pcrs       16    13</span>
<span class="co">#&gt; samples     4     8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $dataset_statistics</span>
<span class="co">#&gt;         nb_reads nb_motus avg_reads sd_reads avg_motus sd_motus</span>
<span class="co">#&gt; pcrs      154553     3182  9659.562 7233.992  537.1875 157.0292</span>
<span class="co">#&gt; samples   154553     3182  9659.562 7233.992  537.1875 157.0292</span></pre></body></html></div>
<p>Note that subsetting of a <code>metabarlist</code> object can be done with <code>subset_metabarlist</code> with any criterion, based either on MOTUs, <em>pcrs</em>, or <em>samples</em> characteristics.</p>
<p>Now lets construct the rarefaction curves with the <code>hill_rarefaction</code> function. The number of sequencing depths used to build these curves is defined by the <code>nsteps</code> argument. For each sequencing depth, the <code>reads</code> table is resampled randomly <code>nboot</code> times so that to obtain an estimate of <span class="math inline">\(^{q}D\)</span> and <span class="math inline">\(Coverage\)</span>. <code>nboot</code> is low in the example below to limit the computing time.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">soil_euk_h20.raref</span> <span class="kw">=</span> <span class="fu"><a href="../reference/hill_rarefaction.html">hill_rarefaction</a></span>(<span class="no">soil_euk_h20</span>, <span class="kw">nboot</span> <span class="kw">=</span> <span class="fl">20</span>, <span class="kw">nsteps</span> <span class="kw">=</span> <span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">soil_euk_h20.raref</span>$<span class="no">hill_table</span>)
<span class="co">#&gt;      pcr_id reads     D0     D0.sd        D1    D1.sd        D2     D2.sd</span>
<span class="co">#&gt; 1 H20-As_r1     5   4.60 0.5026247  4.475125 1.149535  4.428571 0.7180353</span>
<span class="co">#&gt; 2 H20-As_r1    10   8.85 0.5871429  8.459690 1.099164  8.068452 1.0261737</span>
<span class="co">#&gt; 3 H20-As_r1   100  52.95 3.4255234 38.532887 1.093537 27.206227 3.5264966</span>
<span class="co">#&gt; 4 H20-As_r1   200  81.00 6.4236405 49.882751 1.094447 31.200148 2.6579229</span>
<span class="co">#&gt; 5 H20-As_r1   500 137.15 5.7699311 63.605358 1.057316 34.475944 2.6283438</span>
<span class="co">#&gt; 6 H20-As_r1  1000 200.25 7.4824356 72.114363 1.051230 34.699759 2.3687740</span>
<span class="co">#&gt;   coverage coverage.sd</span>
<span class="co">#&gt; 1   0.1600 0.201049876</span>
<span class="co">#&gt; 2   0.2150 0.098808693</span>
<span class="co">#&gt; 3   0.6420 0.042993268</span>
<span class="co">#&gt; 4   0.7525 0.036796024</span>
<span class="co">#&gt; 5   0.8468 0.015524514</span>
<span class="co">#&gt; 6   0.8980 0.009531112</span></pre></body></html></div>
<p>The <code>hill_rarefaction</code> function produces an object with several elements. The most relevant one is the <code>hill_table</code> a table indicating the <code>pcr_id</code>, the sequencing depth at which the <em>pcr</em> was resampled, and the corresponding diversity / coverage indices. The output of <code>hill_rarefaction</code> can be used to draw rarefaction curves.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/hill_rarefaction.html">gghill_rarefaction</a></span>(<span class="no">soil_euk_h20.raref</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/gghill-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The user may also want to distinguish different types of <em>samples</em> on these curves, for example here soil vs. litter <em>samples</em>. This can be achieved as follows:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Define a vector containing the Material info for each pcrs </span>
<span class="no">material</span> <span class="kw">&lt;-</span> <span class="no">soil_euk_h20</span>$<span class="no">samples</span>$<span class="no">Material</span>[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="no">soil_euk_h20</span>$<span class="no">pcrs</span>$<span class="no">sample_id</span>,
                                                <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk_h20</span>$<span class="no">samples</span>))]

<span class="co"># Use of gghill_rarefaction requires a vector with named pcrs</span>
<span class="no">material</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="no">material</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk_h20</span>$<span class="no">pcrs</span>))

<span class="co"># Plot</span>
<span class="no">p</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/hill_rarefaction.html">gghill_rarefaction</a></span>(<span class="no">soil_euk_h20.raref</span>, <span class="kw">group</span><span class="kw">=</span><span class="no">material</span>)
<span class="no">p</span> + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"goldenrod4"</span>, <span class="st">"brown4"</span>, <span class="st">"grey"</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"goldenrod4"</span>, <span class="st">"brown4"</span>, <span class="st">"grey"</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">color</span><span class="kw">=</span><span class="st">"Material type"</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/gghill2-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The obtained curves tell us a couple of things about the <em>pcrs</em>:<br>
- The sampling coverage is relatively high even at intermediate sequencing depths (but see remarks above).<br>
- The less weight is given to rare MOTUs (i.e. the higher <span class="math inline">\(q\)</span> is), the better <span class="math inline">\(^{q}D\)</span> is estimated.<br>
- Litter <em>samples</em> in general tend to be less diverse than soil <em>samples</em>.</p>
</div>
</div>
<div id="flagging-spurious-signal" class="section level2">
<h2 class="hasAnchor">
<a href="#flagging-spurious-signal" class="anchor"></a>Flagging spurious signal</h2>
<p>The next steps of the analysis usually consist in identifying potential spurious signal, whether in terms of MOTUs (e.g. contaminants, untargeted taxa due to primer lack of specificity), MOTUs abundances (i.e. tag-jumps ), or <em>pcrs</em> (<em>pcrs</em> yielding low amounts of reads or with low replicability). In this tutorial, the flagging is done by creating a boolean vector (i.e. <code>TRUE/FALSE</code> where <code>TRUE</code> is good quality signal) specific to each flagging criterion.</p>
<div id="detecting-contaminants-based-on-experimental-negative-controls" class="section level3">
<h3 class="hasAnchor">
<a href="#detecting-contaminants-based-on-experimental-negative-controls" class="anchor"></a>Detecting contaminants based on experimental negative controls</h3>
<p>Contamination can occur at multiple stages, from sample collection to library preparation. Detecting these contaminants can be done through the use of negative controls at each stage <span class="citation">(reviewed in P. Taberlet et al. 2018; Zinger et al. 2019)</span>, as is the case for the <code>soil_euk</code> dataset.</p>
<p>Due to the tagjump bias, many genuine MOTUs that are abundant in <em>samples</em> can be detected in negative controls. Consequently, simply removing from the dataset any MOTU that occurs in negative controls is a very bad idea.</p>
<p>A contaminant should be preferentially amplified in negative controls since there is no competing DNA. The function <code>contaslayer</code> relies on this assumption and detects MOTUs whose relative abundance across the whole dataset is highest in negative controls. Note however that this approach won’t be appropriate if the negative controls have been contaminated with biological samples. In this case,<code>contaslayer</code> should identify MOTUs that are dominants in <em>samples</em>.</p>
<p>The function <code>contaslayer</code> adds a new column in table <code>motus</code> indicating whether the MOTU is a genuine MOTU (<code>TRUE</code>) or a contaminant (<code>FALSE</code>). Below, this is demonstrated with the detection of contaminants from the DNA extraction step.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Identifying extraction contaminants</span>
<span class="no">soil_euk</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/contaslayer.html">contaslayer</a></span>(<span class="no">soil_euk</span>,
                         <span class="kw">control_types</span> <span class="kw">=</span> <span class="st">"extraction"</span>,
                         <span class="kw">output_col</span> <span class="kw">=</span> <span class="st">"not_an_extraction_conta"</span>)

<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">not_an_extraction_conta</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;    66 12581</span></pre></body></html></div>
<p><code>contaslayer</code> identified here 66 contaminant MOTUs. Below are the ten most common extraction contaminants identified by <code>contaslayer</code> in the <code>soil_euk</code> dataset.</p>
<table class="table table table-striped table-hover table-condensed">
<thead><tr>
<th style="text-align:right;">
total # reads
</th>
<th style="text-align:right;">
similarity to ref DB
</th>
<th style="text-align:left;">
full taxonomic path
</th>
<th style="text-align:left;">
sequence
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
195920
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root, Eukaryota
</td>
<td style="text-align:left;">
ctcaaacttccatcggcttgagccgatagtccctctaagaagccggcgaccagccaaagctagcctggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:right;">
23499
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root, Eukaryota, Viridiplantae, Streptophyta, Streptophytina, Embryophyta, Tracheophyta, Euphyllophyta, Spermatophyta, Magnoliophyta, Mesangiospermae, eudicotyledons, Gunneridae, Pentapetalae, rosids, malvids, Sapindales
</td>
<td style="text-align:left;">
ctcaaacttccttggcctaagcggccatagtccctctaagaagctggccacggagggatacctccgcatagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:right;">
8296
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root, Eukaryota, Opisthokonta, Fungi, Dikarya, Ascomycota, saccharomyceta, Saccharomycotina, Saccharomycetes, Saccharomycetales
</td>
<td style="text-align:left;">
ctcaaacttccatcgacttgaaatcgatagtccctctaagaagtgactataccagcaaaagctagcagcactatttagtaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:right;">
4010
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root
</td>
<td style="text-align:left;">
ctcaaacttccatcggcttgaaaccgatagtccctctaagaagtggataaccagcaaatgctagcaccactatttagtaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:right;">
1055
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root, Eukaryota, Opisthokonta, Fungi, Dikarya, Basidiomycota, Ustilaginomycotina, Exobasidiomycetes, Entylomatales
</td>
<td style="text-align:left;">
ctcaaacttccattagctaaacgccaatagtccctctaagaagccagcggctaaccatagtcggccgggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:right;">
966
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root, Eukaryota, Stramenopiles, Chrysophyceae, Chromulinales, Chromulinaceae
</td>
<td style="text-align:left;">
ccccaacttcctttggttagtcaccaaaagtccctctaagaagcttacgtcaatactagtgcattaacaaaactatttagcaggcgggggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:right;">
611
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root, Eukaryota, Opisthokonta, Fungi, Dikarya, Basidiomycota, Agaricomycotina, Tremellomycetes
</td>
<td style="text-align:left;">
ctcaaacttccaacagctaaacgctgttagtccctctaagaagacagcggccagcaaaagccgaccggtctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:right;">
473
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root, Eukaryota, Viridiplantae, Streptophyta, Streptophytina, Embryophyta, Tracheophyta, Euphyllophyta, Spermatophyta, Magnoliophyta, Mesangiospermae, Liliopsida, Petrosaviidae, commelinids, Poales, Poaceae, BOP clade, Pooideae
</td>
<td style="text-align:left;">
ctcaaacttccgtcgcctaaacggcgatagtccctctaagaagctagctgcggagggatggctccgcatagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:right;">
456
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root, Eukaryota, Opisthokonta, Fungi, Dikarya, Basidiomycota, Ustilaginomycotina, Malasseziomycetes, Malasseziales, Malasseziaceae, Malassezia
</td>
<td style="text-align:left;">
ctcaaacttccattggctaaacgccaatagtccctctaagaagccagcagccaaccatagtcgactgggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:right;">
189
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
root, Eukaryota, Viridiplantae, Streptophyta, Streptophytina, Embryophyta, Tracheophyta, Euphyllophyta, Spermatophyta, Magnoliophyta, Mesangiospermae
</td>
<td style="text-align:left;">
ctcaaacttccgtggcctaaacggccatagtccctctaagaagctggccgcggagggatgcctccgtgtagctagttagcaggctgaggtctcgttcgttat
</td>
</tr>
</tbody>
</table>
<p>The identified contaminants correspond to metazoans (including humans), protists, fungi and plants. The most abundant contaminant does not have fine taxonomic identification in the <code>soil_euk</code> dataset, but a BLAST search of the sequence in NCBI suggests that it most likely corresponds to a <em>Fusarium</em>, a notorious laboratory contaminant.</p>
<p>Next, one may want to know how these contamiants are distributed across the <code>soil_euk</code> dataset. For example, the contamination could be present only in one PCR plate for some reason. To assess this, one could use <code>ggpcrplate</code> to plot the distribution of the most common contaminant across the PCR setup:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Identify the most common contaminant</span>
<span class="co"># get contaminant ids</span>
<span class="no">id</span> <span class="kw">&lt;-</span> !<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">not_an_extraction_conta</span>
<span class="no">max.conta</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>[<span class="no">id</span>,])[<span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>[<span class="no">id</span>, <span class="st">"count"</span>])]

<span class="co">#... and its distribution and relative abundance in each pcr</span>
<span class="fu"><a href="../reference/ggpcrplate.html">ggpcrplate</a></span>(<span class="no">soil_euk</span>, <span class="kw">legend_title</span> <span class="kw">=</span> <span class="st">"#reads of most \nabundant contaminant"</span>,
           <span class="kw">FUN</span> <span class="kw">=</span> <span class="kw">function</span>(<span class="no">m</span>) {<span class="no">m</span>$<span class="no">reads</span>[, <span class="no">max.conta</span>]/<span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">m</span>$<span class="no">reads</span>)})</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/contaslayer3-1.png" width="672" style="display: block; margin: auto;"></p>
<p>In the <code>soil_euk</code> dataset, the most abundant contaminant occurs across all samples, but in general at low relative abundance (on average 1.82 %) here.</p>
<p>One can also determine whether <em>pcrs</em> exhibit generally high proportions of contaminants.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Compute relative abundance of all pcr contaminants together </span>
<span class="no">a</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">conta.relab</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">soil_euk</span>$<span class="no">reads</span>[,!<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">not_an_extraction_conta</span>]) /
                                    <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">soil_euk</span>$<span class="no">reads</span>))
<span class="co"># Add information on control types</span>
<span class="no">a</span>$<span class="no">control_type</span> <span class="kw">&lt;-</span> <span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">control_type</span>[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">a</span>), <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>))]

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">a</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">control_type</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">conta.relab</span>, <span class="kw">color</span><span class="kw">=</span><span class="no">control_type</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="kw">alpha</span><span class="kw">=</span><span class="fl">0.5</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"brown"</span>, <span class="st">"red"</span>, <span class="st">"cyan4"</span>,<span class="st">"pink"</span>), <span class="kw">na.value</span> <span class="kw">=</span> <span class="st">"darkgrey"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="kw">NULL</span>, <span class="kw">y</span><span class="kw">=</span><span class="st">"Prop. Reads (log10)"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>()</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/contaslayer4-1.png" width="384" style="display: block; margin: auto;"></p>
<p>Overall, <em>samples</em> yield much less amounts of extraction contaminants than experimental negative controls. Some <em>pcrs</em> corresponding to <em>samples</em> have, however, &gt; 10% of their reads corresponding to contaminants and could be dysfunctional. Based on this criteria <em>pcrs</em> could be flagged as follows for potential downstream filtering:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># flag pcrs with total contaminant relative abundance &gt; 10% of reads)</span>
<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">low_contamination_level</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">a</span>$<span class="no">conta.relab</span>[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>), <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">a</span>))]<span class="kw">&gt;</span><span class="fl">1e-1</span>,  <span class="no">F</span>, <span class="no">T</span>)

<span class="co"># Proportion of potentially functional (TRUE) vs. dysfunctional (FALSE) pcrs</span>
<span class="co"># (controls included) based on this criterion</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">low_contamination_level</span>) / <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;     FALSE      TRUE </span>
<span class="co">#&gt; 0.1666667 0.8046875</span></pre></body></html></div>
<p>Note that this step should be done for all types of contaminants (coming from extraction, pcr or sequencing). The user can identify each of them separetly or without differentiating the different controls types.</p>
</div>
<div id="flagging-spurious-or-non-target-motus" class="section level3">
<h3 class="hasAnchor">
<a href="#flagging-spurious-or-non-target-motus" class="anchor"></a>Flagging spurious or non-target MOTUs</h3>
<p>Non-target sequences can be amplified if the primers are not specific enough. On the other hand, some highly degraded sequences can be produced throughout the data production process, such as primer dimers, or chimeras from multiple parents (hereafter referred to as spurious MOTUs). To detect these, one can use the information related to taxonomic assignments and associated similarity scores.</p>
<p>Since the <code>soil_euk</code> dataset was obtained with primers that target eukaryotes, non-eukaryotic MOTUs should be excluded. At this stage of the analysis, we only flag MOTUs based on this criterion.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co">#Flag MOTUs corresponding to target (TRUE) vs. non-target (FALSE) taxa </span>
<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">target_taxon</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"Eukaryota"</span>, <span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">path</span>)

<span class="co"># Proportion of each of these over total number of MOTUs</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">target_taxon</span>) / <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;      FALSE       TRUE </span>
<span class="co">#&gt; 0.04356764 0.95643236</span>

<span class="co"># Intersection with extraction contaminant flags (not contaminant = T)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">target_taxon</span>,
      <span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">not_an_extraction_conta</span>)
<span class="co">#&gt;        </span>
<span class="co">#&gt;         FALSE  TRUE</span>
<span class="co">#&gt;   FALSE     8   543</span>
<span class="co">#&gt;   TRUE     58 12038</span></pre></body></html></div>
<p>Here we flagged 8 MOTUS as non-targets, which were already flagged as potential extraction contaminants.</p>
<p>Next, we want to identify MOTUs whose sequence is too dissimilar from references. This filtering criterion relies on the assumption that current reference databases capture most of the diversity at broad taxonomic levels (i.e. already have for example at least one representative of each phyla). Considering this, MOTUs being too distant from reference databases are more likely to be a degraded sequence, especially if such MOTUs are relatively numerous and of low abundance. To assess this, one can use the distribution of MOTU similarity scores, weighted and unweighted by their relative abundance.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Plot the unweighted distribution of MOTUs similarity scores </span>
<span class="no">a</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">best_identity.order_filtered_embl_r136_noenv_EUK</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="kw">color</span><span class="kw">=</span><span class="st">"grey"</span>, <span class="kw">fill</span><span class="kw">=</span><span class="st">"white"</span>, <span class="kw">bins</span><span class="kw">=</span><span class="fl">20</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="kw">xintercept</span> <span class="kw">=</span> <span class="fl">0.8</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"orange"</span>, <span class="kw">lty</span><span class="kw">=</span><span class="fl">2</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">panel.grid</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>()) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="st">"% similarity against best match"</span>, <span class="kw">y</span><span class="kw">=</span><span class="st">"# OTUs"</span>)

<span class="co"># Same for the weighted distribution</span>
<span class="no">b</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>,
         <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">best_identity.order_filtered_embl_r136_noenv_EUK</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">..count..</span>, <span class="kw">weight</span> <span class="kw">=</span> <span class="no">count</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="kw">color</span><span class="kw">=</span><span class="st">"grey"</span>, <span class="kw">fill</span><span class="kw">=</span><span class="st">"white"</span>, <span class="kw">bins</span><span class="kw">=</span><span class="fl">20</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="kw">xintercept</span> <span class="kw">=</span> <span class="fl">0.8</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"orange"</span>, <span class="kw">lty</span><span class="kw">=</span><span class="fl">2</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">panel.grid</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>()) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="st">"% similarity against best match"</span>, <span class="kw">y</span><span class="kw">=</span><span class="st">"# Reads"</span>)

<span class="co"># Combine plots into one</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">cowplot</span>)
<span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/ggdraw.html">ggdraw</a></span>() +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/draw_plot.html">draw_plot</a></span>(<span class="no">a</span>, <span class="kw">x</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">y</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">width</span> <span class="kw">=</span> <span class="fl">0.5</span>) +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/draw_plot.html">draw_plot</a></span>(<span class="no">b</span>, <span class="kw">x</span><span class="kw">=</span><span class="fl">0.5</span>, <span class="kw">y</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">width</span> <span class="kw">=</span> <span class="fl">0.5</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/nontarget3-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Here we may consider any MOTU as degraded sequences if its sequence similarity is &lt; 80% against its best match in the reference database.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Flag not degraded (TRUE) vs. potentially degraded sequences (FALSE)</span>
<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">not_degraded</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">best_identity.order_filtered_embl_r136_noenv_EUK</span> <span class="kw">&lt;</span> <span class="fl">0.8</span>, <span class="no">F</span>, <span class="no">T</span>)

<span class="co"># Proportion of each of these over total number of MOTUs</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">not_degraded</span>) / <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;    FALSE     TRUE </span>
<span class="co">#&gt; 0.111568 0.888432</span>

<span class="co"># Intersection with other flags</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">target_taxon</span>,
      <span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">not_an_extraction_conta</span>,
      <span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">not_degraded</span>)
<span class="co">#&gt; , ,  = FALSE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;        </span>
<span class="co">#&gt;         FALSE  TRUE</span>
<span class="co">#&gt;   FALSE     7   438</span>
<span class="co">#&gt;   TRUE     10   956</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; , ,  = TRUE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;        </span>
<span class="co">#&gt;         FALSE  TRUE</span>
<span class="co">#&gt;   FALSE     1   105</span>
<span class="co">#&gt;   TRUE     48 11082</span></pre></body></html></div>
<p>Here, we find that 7 non-target MOTUs are also extraction contaminants and potentially degraded fragments.</p>
</div>
<div id="detecting-pcr-outliers" class="section level3">
<h3 class="hasAnchor">
<a href="#detecting-pcr-outliers" class="anchor"></a>Detecting PCR outliers</h3>
<p>A first way to identify dysfunctional PCRs is to flag them based on the <em>pcr</em> sequencing depth</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">nb_reads</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="kw">bins</span><span class="kw">=</span><span class="fl">40</span>, <span class="kw">color</span><span class="kw">=</span><span class="st">"grey"</span>, <span class="kw">fill</span><span class="kw">=</span><span class="st">"white"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="kw">xintercept</span> <span class="kw">=</span> <span class="fl">1e3</span>, <span class="kw">lty</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">color</span><span class="kw">=</span><span class="st">"orange"</span>) + <span class="co"># threshold</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="st">"# Reads (with all OTUs and PCRs)"</span>,
        <span class="kw">y</span><span class="kw">=</span><span class="st">"# PCRs"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">panel.grid</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>())</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/dysfuncPCR1-1.png" width="480" style="display: block; margin: auto;"></p>
<p>The histogram above includes negative controls, which - fortunately - yield low amounts of reads in general, but also <em>pcrs</em> obtained from <em>samples</em>, as shown below.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Flag pcrs with an acceptable sequencing depth (TRUE) or inacceptable one (FALSE)</span>
<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">seqdepth_ok</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">nb_reads</span> <span class="kw">&lt;</span> <span class="fl">1e3</span>, <span class="no">F</span>, <span class="no">T</span>)

<span class="co"># Proportion of each of these over total number of pcrs, control excluded</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">seqdepth_ok</span>[<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">type</span><span class="kw">==</span><span class="st">"sample"</span>]) /
  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>[<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">type</span><span class="kw">==</span><span class="st">"sample"</span>,])
<span class="co">#&gt; </span>
<span class="co">#&gt;      FALSE       TRUE </span>
<span class="co">#&gt; 0.02734375 0.97265625</span></pre></body></html></div>
<p>A second way to evaluate the PCR quality is to assess their reproducibility. The <code>pcrslayer</code> and associated functions (e.g. <code>pcr_within_between</code>) enable the detection of outlier replicates within <em>pcrs</em> using different methods, with a common underlying assumption: the compositional dissimilarities in MOTUs between replicate <em>pcrs</em> from a same <em>sample</em> should be lower than that observed between <em>pcrs</em> obtained from different <em>samples</em> (see help page for more details). This assessment is obviously only possible when the experimental designs includes PCR replicates for each biological samples. For these function to run, pcrs yielding no reads, as well as negative controls, should be excluded.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Subsetting the metabarlist</span>
<span class="no">soil_euk_sub</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/subset_metabarlist.html">subset_metabarlist</a></span>(<span class="no">soil_euk</span>,
                                   <span class="kw">table</span> <span class="kw">=</span> <span class="st">"pcrs"</span>,
                                   <span class="kw">indices</span> <span class="kw">=</span> <span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">nb_reads</span><span class="kw">&gt;</span><span class="fl">0</span> <span class="kw">&amp;</span> (
                                     <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">control_type</span>) <span class="kw">|</span>
                                       <span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">control_type</span><span class="kw">==</span><span class="st">"positive"</span>))

<span class="co"># First visualization</span>
<span class="no">comp1</span> <span class="kw">=</span> <span class="fu"><a href="../reference/pcrslayer.html">pcr_within_between</a></span>(<span class="no">soil_euk_sub</span>)
<span class="fu"><a href="../reference/pcrslayer.html">check_pcr_thresh</a></span>(<span class="no">comp1</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/pcrsalyer1-1.png" width="480" style="display: block; margin: auto;"></p>
<p>In the density plots shown above, one can see that several <em>pcrs</em> replicates are as distant than <em>pcrs</em> from different <em>samples</em>. Let’s now flag <em>pcrs</em> based on this criterion and store this in the <code>output_col</code> column of the <code>pcr</code> table:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Flagging</span>
<span class="no">soil_euk_sub</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/pcrslayer.html">pcrslayer</a></span>(<span class="no">soil_euk_sub</span>, <span class="kw">output_col</span> <span class="kw">=</span> <span class="st">"replicating_pcr"</span>, <span class="kw">plot</span> <span class="kw">=</span> <span class="no">F</span>)
<span class="co">#&gt; [1] "Iteration 1"</span>
<span class="co">#&gt; [1] "Iteration 2"</span>
<span class="co">#&gt; [1] "Iteration 3"</span>

<span class="co"># Proportion of replicating pcrs (TRUE)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">soil_euk_sub</span>$<span class="no">pcrs</span>$<span class="no">replicating_pcr</span>) /
  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">soil_euk_sub</span>$<span class="no">pcrs</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;      FALSE       TRUE </span>
<span class="co">#&gt; 0.01736111 0.98263889</span>

<span class="co"># Intersection with the sequencing depth criterion</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">soil_euk_sub</span>$<span class="no">pcrs</span>$<span class="no">seqdepth_ok</span>,
      <span class="no">soil_euk_sub</span>$<span class="no">pcrs</span>$<span class="no">replicating_pcr</span>)
<span class="co">#&gt;        </span>
<span class="co">#&gt;         FALSE TRUE</span>
<span class="co">#&gt;   FALSE     3    5</span>
<span class="co">#&gt;   TRUE      2  278</span></pre></body></html></div>
<p>Here, 3 <em>pcrs</em> out of the 5 detected as outliers had low sequencing depth. One can also visualise the replicates and outliers behaviour in a Principal Coordinate Analysis with the function <code>check_pcr_repl</code>.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Distinguish between pcrs obtained from samples from positive controls</span>
<span class="no">mds</span> <span class="kw">=</span> <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_repl</a></span>(<span class="no">soil_euk_sub</span>,
                     <span class="kw">groups</span> <span class="kw">=</span> <span class="no">soil_euk_sub</span>$<span class="no">pcrs</span>$<span class="no">type</span>,
                     <span class="kw">funcpcr</span> <span class="kw">=</span> <span class="no">soil_euk_sub</span>$<span class="no">pcrs</span>$<span class="no">replicating_pcr</span>)
<span class="no">mds</span> + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="st">"pcr type"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cyan4"</span>, <span class="st">"gray"</span>))</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/pcrsalyer3-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Now report the flagging in the initial <code>metabarlist</code></p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">replicating_pcr</span> <span class="kw">&lt;-</span> <span class="fl">NA</span>
<span class="no">soil_euk</span>$<span class="no">pcrs</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk_sub</span>$<span class="no">pcrs</span>),<span class="st">"replicating_pcr"</span>] <span class="kw">&lt;-</span> <span class="no">soil_euk_sub</span>$<span class="no">pcrs</span>$<span class="no">replicating_pcr</span></pre></body></html></div>
</div>
<div id="lowering-tag-jumps" class="section level3">
<h3 class="hasAnchor">
<a href="#lowering-tag-jumps" class="anchor"></a>Lowering tag-jumps</h3>
<p>Tag-jumps are frequency-dependent, i.e. abundant genuine MOTUs are more likely to be found in low abundance in samples were they are not supposed to be than rare genuine MOTUs. To reduce the amount of such false positives, the function <code>tagjumpslayer</code> considers each MOTU separately and corrects its abundance in <em>pcrs</em> (see <code>tagjumpslayer</code> help for more information on possible correction methods) when the MOTU relative abundance over the entire dataset is below a given threshold. Such data a curation strategy is similar to what has been proposed by <span class="citation">Esling, Lejzerowicz, and Pawlowski (2015)</span>. Effect of this threshold can be evaluated by testing how this filtration procedure affects basic dataset characteristics (e.g. # MOTUs or reads) at different levels, as exemplified below.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Define a vector of thresholds to test</span>
<span class="no">thresholds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>,<span class="fl">1e-4</span>,<span class="fl">1e-3</span>, <span class="fl">1e-2</span>, <span class="fl">3e-2</span>, <span class="fl">5e-2</span>)

<span class="co"># Run the tests and stores the results in a list</span>
<span class="no">tests</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">thresholds</span>, <span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="../reference/tagjumpslayer.html">tagjumpslayer</a></span>(<span class="no">soil_euk</span>,<span class="no">x</span>))
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">tests</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"t_"</span>, <span class="no">thresholds</span>, <span class="kw">sep</span><span class="kw">=</span><span class="st">""</span>)

<span class="co"># Format the data for ggplot with amount of reads at each threshold</span>
<span class="no">tmp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="st">"rbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">tests</span>, <span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">x</span>$<span class="no">reads</span>)))))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tmp</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"threshold"</span>, <span class="st">"sample"</span>, <span class="st">"abundance"</span>)

<span class="co"># Add richness in MOTUs at each threshold</span>
<span class="no">tmp</span>$<span class="no">richness</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="st">"rbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">tests</span>, <span class="kw">function</span>(<span class="no">x</span>) {
    <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">x</span>$<span class="no">reads</span> <span class="kw">&gt;</span> <span class="fl">0</span>)
  }))))$<span class="no">value</span>

<span class="co"># Add control type information on pcrs and make data curation threshold numeric</span>
<span class="no">tmp</span>$<span class="no">controls</span> <span class="kw">&lt;-</span> <span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">control_type</span>[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="no">tmp</span>$<span class="no">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk</span>$<span class="no">pcrs</span>))]
<span class="no">tmp</span>$<span class="no">threshold</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"t_"</span>, <span class="st">""</span>, <span class="no">tmp</span>$<span class="no">threshold</span>))

<span class="co"># New table formatting for ggplot</span>
<span class="no">tmp2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span>(<span class="no">tmp</span>, <span class="kw">id.vars</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tmp</span>)[-<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"abundance|richness"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tmp</span>))])

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">tmp2</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">threshold</span>), <span class="kw">y</span><span class="kw">=</span><span class="no">value</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>(<span class="kw">color</span><span class="kw">=</span><span class="st">"grey40"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="kw">xintercept</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="no">tmp2</span>$<span class="no">threshold</span>)) <span class="kw">==</span> <span class="st">"0.01"</span>), <span class="kw">col</span><span class="kw">=</span><span class="st">"orange"</span>, <span class="kw">lty</span><span class="kw">=</span><span class="fl">2</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span><span class="kw">=</span><span class="no">controls</span>), <span class="kw">width</span> <span class="kw">=</span> <span class="fl">0.2</span>, <span class="kw">alpha</span><span class="kw">=</span><span class="fl">0.5</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"brown"</span>, <span class="st">"red"</span>, <span class="st">"cyan4"</span>,<span class="st">"pink"</span>), <span class="kw">na.value</span> <span class="kw">=</span> <span class="st">"darkgrey"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">variable</span>+<span class="no">controls</span>, <span class="kw">scale</span><span class="kw">=</span><span class="st">"free_y"</span>, <span class="kw">ncol</span><span class="kw">=</span><span class="fl">5</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="st">"MOTU pcr : total abundance filtering threshold"</span>, <span class="kw">y</span><span class="kw">=</span><span class="st">"# Reads/MOTUs"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">panel.grid</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
        <span class="kw">strip.background</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
        <span class="kw">axis.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">angle</span><span class="kw">=</span><span class="fl">40</span>, <span class="kw">h</span><span class="kw">=</span><span class="fl">1</span>),
        <span class="kw">legend.position</span> <span class="kw">=</span> <span class="st">"none"</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/tagjump1-1.png" width="960" style="display: block; margin: auto;"></p>
<p>A threshold of 0.01 leads to a drop in both the number of reads and of MOTUs in sequencing negative controls. This drop is also noticeable in terms of the number of MOTUs in <em>pcrs</em> obtained from other controls as compared to those obtained from <em>samples</em>. The former are expected to be void of environmental MOTUs, and tag-jumps should be more visible/important in these <em>pcrs</em>. Note that this procedure primarily affects MOTU diversity in <em>pcrs</em>, and poorly the number of reads in <em>pcrs</em>.</p>
<p>As for above, <em>pcrs</em> containing large amounts of MOTUs identified as potentially artifactual or where tag-jumps filtering strongly affects the number of reads in <em>pcrs</em> can be flagged as potentially dysfunctional.</p>
</div>
</div>
<div id="summarizing-the-noise-in-the-soil_euk-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#summarizing-the-noise-in-the-soil_euk-dataset" class="anchor"></a>Summarizing the noise in the <code>soil_euk</code> dataset</h2>
<p>We can now get an overview of the amount of noise identified with the criteria used above, for both the number of MOTUs and their associated readcount.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Create a table of MOTUs quality criteria </span>
<span class="co"># noise is identified as FALSE in soil_euk, the "!" transforms it to TRUE</span>
<span class="no">motus.qual</span> <span class="kw">&lt;-</span> !<span class="no">soil_euk</span>$<span class="no">motus</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"not_an_extraction_conta"</span>, <span class="st">"target_taxon"</span>, <span class="st">"not_degraded"</span>)]
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">motus.qual</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"extraction_conta"</span>, <span class="st">"untargeted_taxon"</span>, <span class="st">"degraded_seq"</span>)

<span class="co"># Proportion of MOTUs potentially artifactual (TRUE) based on the criteria used</span>
<span class="fu"><a href="https://rdrr.io/r/base/prop.table.html">prop.table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">motus.qual</span>, <span class="fl">1</span>, <span class="no">sum</span>) <span class="kw">&gt;</span> <span class="fl">0</span>))
<span class="co">#&gt; </span>
<span class="co">#&gt;     FALSE      TRUE </span>
<span class="co">#&gt; 0.8762552 0.1237448</span>

<span class="co"># Corresponding proportion of artifactual reads (TRUE)</span>
<span class="fu"><a href="https://rdrr.io/r/base/prop.table.html">prop.table</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/xtabs.html">xtabs</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">count</span>~<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">motus.qual</span>, <span class="fl">1</span>, <span class="no">sum</span>) <span class="kw">&gt;</span> <span class="fl">0</span>))
<span class="co">#&gt; apply(motus.qual, 1, sum) &gt; 0</span>
<span class="co">#&gt;      FALSE       TRUE </span>
<span class="co">#&gt; 0.90582023 0.09417977</span>

<span class="co"># Proportion of MOTUs and reads potentially artifactual for each criterion</span>
<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">motus.qual</span>, <span class="fl">2</span>, <span class="no">sum</span>) / <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">motus.qual</span>)
<span class="co">#&gt; extraction_conta untargeted_taxon     degraded_seq </span>
<span class="co">#&gt;      0.005218629      0.043567645      0.111567961</span>
<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">motus.qual</span>, <span class="fl">2</span>, <span class="kw">function</span>(<span class="no">x</span>) <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">count</span>[<span class="no">x</span>])/<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">soil_euk</span>$<span class="no">motus</span>$<span class="no">count</span>))
<span class="co">#&gt; extraction_conta untargeted_taxon     degraded_seq </span>
<span class="co">#&gt;       0.06669166       0.01372710       0.02629960</span>

<span class="no">tmp.motus</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">motus.qual</span>), <span class="kw">function</span>(<span class="no">x</span>) {
    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">motus.qual</span>[,<span class="no">x</span>]<span class="kw">==</span><span class="no">T</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">motus.qual</span>)[<span class="no">x</span>], <span class="fl">NA</span>)}), <span class="fl">1</span>, <span class="kw">function</span>(<span class="no">x</span>) {
      <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">x</span>)), <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">"|"</span>)
      })
<span class="no">tmp.motus</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"^$"</span>, <span class="st">"not_artefactual"</span>, <span class="no">tmp.motus</span>))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tmp.motus</span>) <span class="kw">&lt;-</span>  <span class="st">"artefact_type"</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">tmp.motus</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">fill</span><span class="kw">=</span><span class="no">artefact_type</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>() +  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span>(<span class="fl">0</span>, <span class="fl">2</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">fill</span><span class="kw">=</span><span class="st">"Artifact type"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_polar.html">coord_polar</a></span>(<span class="kw">theta</span><span class="kw">=</span><span class="st">"y"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="kw">palette</span> <span class="kw">=</span> <span class="st">"Set3"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">legend.direction</span> <span class="kw">=</span> <span class="st">"vertical"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"MOTUs artefacts overview"</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/noiseMOTUs-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The above shows that MOTUs flagged as potentially artefactual account for ca. 10% of the dataset’s diversity and roughly the same in terms of readcount. Most of these artifact MOTUs are rare and correspond to sequences which are potentially highly degraded, with very low sequence similarity against the EMBL reference database. The most abundant artifacts MOTUs were identified as contaminants.</p>
<p>Let’s do the same with <em>pcrs</em></p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Create a table of pcrs quality criteria </span>
<span class="co"># noise is identified as FALSE in soil_euk, the "!" transforms it to TRUE</span>
<span class="no">pcrs.qual</span> <span class="kw">&lt;-</span> !<span class="no">soil_euk</span>$<span class="no">pcrs</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"low_contamination_level"</span>, <span class="st">"seqdepth_ok"</span>, <span class="st">"replicating_pcr"</span>)]
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">pcrs.qual</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"high_contamination_level"</span>, <span class="st">"low_seqdepth"</span>, <span class="st">"outliers"</span>)

<span class="co"># Proportion of pcrs potentially artifactual (TRUE) based on the criteria used</span>
<span class="co"># excluding controls</span>
<span class="fu"><a href="https://rdrr.io/r/base/prop.table.html">prop.table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">pcrs.qual</span>[<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">type</span><span class="kw">==</span><span class="st">"sample"</span>,], <span class="fl">1</span>, <span class="no">sum</span>) <span class="kw">&gt;</span> <span class="fl">0</span>))
<span class="co">#&gt; </span>
<span class="co">#&gt;     FALSE      TRUE </span>
<span class="co">#&gt; 0.8671875 0.1328125</span>

<span class="co"># Proportion of MOTUs and reads potentially artifactual for each criterion</span>
<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">pcrs.qual</span>[<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">type</span><span class="kw">==</span><span class="st">"sample"</span>,], <span class="fl">2</span>, <span class="no">sum</span>) / <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">pcrs.qual</span>[<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">type</span><span class="kw">==</span><span class="st">"sample"</span>,])
<span class="co">#&gt; high_contamination_level             low_seqdepth                 outliers </span>
<span class="co">#&gt;               0.10937500               0.02734375               0.01562500</span>

<span class="no">tmp.pcrs</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">pcrs.qual</span>), <span class="kw">function</span>(<span class="no">x</span>) {
    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">pcrs.qual</span>[<span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">type</span><span class="kw">==</span><span class="st">"sample"</span>,<span class="no">x</span>]<span class="kw">==</span><span class="no">T</span>,
           <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">pcrs.qual</span>)[<span class="no">x</span>], <span class="fl">NA</span>)}), <span class="fl">1</span>, <span class="kw">function</span>(<span class="no">x</span>) {
      <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">x</span>)), <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">"|"</span>)
      })
<span class="no">tmp.pcrs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"^$"</span>, <span class="st">"not_artefactual"</span>, <span class="no">tmp.pcrs</span>))

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tmp.pcrs</span>) <span class="kw">&lt;-</span> <span class="st">"artefact_type"</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">tmp.pcrs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">fill</span><span class="kw">=</span><span class="no">artefact_type</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>() +  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span>(<span class="fl">0</span>, <span class="fl">2</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">fill</span><span class="kw">=</span><span class="st">"Artifact type"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_polar.html">coord_polar</a></span>(<span class="kw">theta</span><span class="kw">=</span><span class="st">"y"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="kw">palette</span> <span class="kw">=</span> <span class="st">"Set3"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">legend.direction</span> <span class="kw">=</span> <span class="st">"vertical"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"PCR artefacts overview"</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/noisePCRs-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The <em>pcrs</em> flagged as potentially dysfunctional account for ca. 10% of the <em>pcrs</em>, most of these corresponding to <em>pcrs</em> yiedling a high amount of contaminants.</p>
</div>
<div id="data-cleaning-and-aggregation" class="section level2">
<h2 class="hasAnchor">
<a href="#data-cleaning-and-aggregation" class="anchor"></a>Data cleaning and aggregation</h2>
<p>The final stage of the analysis consists in removing data based on the criteria defined above and aggregating <em>pcrs</em> to get rid of technical replicates. Here, the decision of which of these flagged criteria to use rests with the user, depending on how they feel the tagged errors impact on the characteristics of their dataset.</p>
<div id="removing-spurious-signal" class="section level3">
<h3 class="hasAnchor">
<a href="#removing-spurious-signal" class="anchor"></a>Removing spurious signal</h3>
<p>First, we will remove suprious MOTUs, PCRs and adjusting read counts by removing tag-jumps . At this stage of the analysis, controls are no longer necessary, and so can also be removed from the dataset.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Use tag-jump corrected metabarlist with the threshold identified above</span>
<span class="no">tmp</span> <span class="kw">&lt;-</span> <span class="no">tests</span><span class="kw">[[</span><span class="st">"t_0.01"</span>]]

<span class="co"># Subset on MOTUs: we keep motus that are defined as TRUE following the </span>
<span class="co"># three criteria below (sum of three TRUE is equal to 3 with the rowSums function)</span>
<span class="no">tmp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/subset_metabarlist.html">subset_metabarlist</a></span>(<span class="no">tmp</span>, <span class="st">"motus"</span>,
                          <span class="kw">indices</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">tmp</span>$<span class="no">motus</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"not_an_extraction_conta"</span>, <span class="st">"target_taxon"</span>,
                                                 <span class="st">"not_degraded"</span>)]) <span class="kw">==</span> <span class="fl">3</span>)
<span class="fu"><a href="../reference/summary_metabarlist.html">summary_metabarlist</a></span>(<span class="no">tmp</span>)
<span class="co">#&gt; $dataset_dimension</span>
<span class="co">#&gt;         n_row n_col</span>
<span class="co">#&gt; reads     384 11082</span>
<span class="co">#&gt; motus   11082    18</span>
<span class="co">#&gt; pcrs      384    16</span>
<span class="co">#&gt; samples    64     8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $dataset_statistics</span>
<span class="co">#&gt;         nb_reads nb_motus avg_reads sd_reads avg_motus sd_motus</span>
<span class="co">#&gt; pcrs     3173438    11082  8264.161 9543.057  285.3333 266.7837</span>
<span class="co">#&gt; samples  2514362    10886  9821.727 9453.080  420.4492 227.2386</span></pre></body></html></div>
<p>Same with <em>pcrs</em></p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Subset on pcrs and keep only controls </span>
<span class="no">soil_euk_clean</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/subset_metabarlist.html">subset_metabarlist</a></span>(<span class="no">tmp</span>, <span class="st">"pcrs"</span>,
                          <span class="kw">indices</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">tmp</span>$<span class="no">pcrs</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"low_contamination_level"</span>,
                                                         <span class="st">"seqdepth_ok"</span>, <span class="st">"replicating_pcr"</span>)]) <span class="kw">==</span> <span class="fl">3</span> <span class="kw">&amp;</span>
                                    <span class="no">tmp</span>$<span class="no">pcrs</span>$<span class="no">type</span> <span class="kw">==</span> <span class="st">"sample"</span>)
<span class="fu"><a href="../reference/summary_metabarlist.html">summary_metabarlist</a></span>(<span class="no">soil_euk_clean</span>)
<span class="co">#&gt; $dataset_dimension</span>
<span class="co">#&gt;         n_row n_col</span>
<span class="co">#&gt; reads     222 10570</span>
<span class="co">#&gt; motus   10570    18</span>
<span class="co">#&gt; pcrs      222    16</span>
<span class="co">#&gt; samples    61     8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $dataset_statistics</span>
<span class="co">#&gt;         nb_reads nb_motus avg_reads sd_reads avg_motus sd_motus</span>
<span class="co">#&gt; pcrs     2193430    10570  9880.315  9540.84  446.3333 228.1212</span>
<span class="co">#&gt; samples  2193430    10570  9880.315  9540.84  446.3333 228.1212</span></pre></body></html></div>
<p>Now check if previous subsetting leads to any empty <em>pcrs</em> or MOTUs</p>
<div class="sourceCode"><html><body><pre class="r"><span class="kw">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">reads</span>)<span class="kw">==</span><span class="fl">0</span>)<span class="kw">&gt;</span><span class="fl">0</span>){<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="st">"empty motus present"</span>)}
<span class="kw">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">reads</span>)<span class="kw">==</span><span class="fl">0</span>)<span class="kw">&gt;</span><span class="fl">0</span>){<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="st">"empty pcrs present"</span>)}</pre></body></html></div>
<p>Since we have now removed certain OTUs or reduced their read counts, we need to update some parameters in the metabarlist (e.g. counts, etc.).</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">soil_euk_clean</span>$<span class="no">motus</span>$<span class="no">count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">reads</span>)
<span class="no">soil_euk_clean</span>$<span class="no">pcrs</span>$<span class="no">nb_reads_postmetabaR</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">reads</span>)
<span class="no">soil_euk_clean</span>$<span class="no">pcrs</span>$<span class="no">nb_motus_postmetabaR</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">reads</span><span class="kw">&gt;</span><span class="fl">0</span>, <span class="no">T</span>, <span class="no">F</span>))</pre></body></html></div>
<p>We can now compare some basic characteristics of the <code>soil_euk</code> before and after data curation with <code>metabaR</code>.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">check</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">pcrs</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"nb_reads"</span>, <span class="st">"nb_reads_postmetabaR"</span>,
                               <span class="st">"nb_motus"</span>, <span class="st">"nb_motus_postmetabaR"</span>)])
<span class="no">check</span>$<span class="no">type</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"motus"</span>, <span class="no">check</span>$<span class="no">variable</span>), <span class="st">"richness"</span>, <span class="st">"abundance"</span>)

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="no">check</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">variable</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">value</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>( <span class="kw">color</span> <span class="kw">=</span> <span class="st">"darkgrey"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="kw">alpha</span><span class="kw">=</span><span class="fl">0.1</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="st">"darkgrey"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">type</span>, <span class="kw">scales</span> <span class="kw">=</span> <span class="st">"free"</span>, <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">5</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">axis.text.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">angle</span><span class="kw">=</span><span class="fl">45</span>, <span class="kw">h</span><span class="kw">=</span><span class="fl">1</span>))</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/beforeafter1-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The sequencing depth and richness of <em>pcrs</em> are not greatly affected by the trimming.<br>
Let’s now see if the signal is changed in terms of beta diversity:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Get row data only for samples</span>
<span class="no">tmp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/subset_metabarlist.html">subset_metabarlist</a></span>(<span class="no">soil_euk</span>, <span class="kw">table</span> <span class="kw">=</span> <span class="st">"pcrs"</span>,
                          <span class="kw">indices</span> <span class="kw">=</span> <span class="no">soil_euk</span>$<span class="no">pcrs</span>$<span class="no">type</span> <span class="kw">==</span> <span class="st">"sample"</span>)

<span class="co"># Add sample biological information for checks</span>
<span class="no">tmp</span>$<span class="no">pcrs</span>$<span class="no">Material</span> <span class="kw">&lt;-</span> <span class="no">tmp</span>$<span class="no">samples</span>$<span class="no">Material</span>[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="no">tmp</span>$<span class="no">pcrs</span>$<span class="no">sample_id</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">tmp</span>$<span class="no">samples</span>))]
<span class="no">tmp</span>$<span class="no">pcrs</span>$<span class="no">Habitat</span> <span class="kw">&lt;-</span> <span class="no">tmp</span>$<span class="no">samples</span>$<span class="no">Habitat</span>[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="no">tmp</span>$<span class="no">pcrs</span>$<span class="no">sample_id</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">tmp</span>$<span class="no">samples</span>))]

<span class="no">soil_euk_clean</span>$<span class="no">pcrs</span>$<span class="no">Material</span> <span class="kw">&lt;-</span>
  <span class="no">soil_euk_clean</span>$<span class="no">samples</span>$<span class="no">Material</span>[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">pcrs</span>$<span class="no">sample_id</span>,
                                        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">samples</span>))]
<span class="no">soil_euk_clean</span>$<span class="no">pcrs</span>$<span class="no">Habitat</span> <span class="kw">&lt;-</span>
  <span class="no">soil_euk_clean</span>$<span class="no">samples</span>$<span class="no">Habitat</span>[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">pcrs</span>$<span class="no">sample_id</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk_clean</span>$<span class="no">samples</span>))]

<span class="co"># Build PCoA ordinations </span>
<span class="no">mds1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_repl</a></span>(<span class="no">tmp</span>,
                      <span class="kw">groups</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="no">tmp</span>$<span class="no">pcrs</span>$<span class="no">Habitat</span>, <span class="no">tmp</span>$<span class="no">pcrs</span>$<span class="no">Material</span>, <span class="kw">sep</span> <span class="kw">=</span> <span class="st">" | "</span>))
<span class="no">mds2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_repl</a></span>(<span class="no">soil_euk_clean</span>,
                       <span class="kw">groups</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(
                         <span class="no">soil_euk_clean</span>$<span class="no">pcrs</span>$<span class="no">Habitat</span>,
                         <span class="no">soil_euk_clean</span>$<span class="no">pcrs</span>$<span class="no">Material</span>,
                         <span class="kw">sep</span> <span class="kw">=</span> <span class="st">" | "</span>))

<span class="co"># Custom colors</span>
<span class="no">a</span> <span class="kw">&lt;-</span> <span class="no">mds1</span> + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="st">"Habitat | Material"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"brown4"</span>, <span class="st">"brown1"</span>, <span class="st">"goldenrod4"</span>, <span class="st">"goldenrod1"</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">legend.position</span> <span class="kw">=</span> <span class="st">"none"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Raw data"</span>)
<span class="no">b</span> <span class="kw">&lt;-</span> <span class="no">mds2</span> + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="st">"Habitat | Material"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"brown4"</span>, <span class="st">"brown1"</span>, <span class="st">"goldenrod4"</span>, <span class="st">"goldenrod1"</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Clean data"</span>)

<span class="co"># Assemble plots</span>
<span class="no">leg</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/get_legend.html">get_legend</a></span>(<span class="no">b</span> + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="kw">shape</span><span class="kw">=</span><span class="no">F</span>) +
                    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">legend.position</span> <span class="kw">=</span> <span class="st">"right"</span>,
                          <span class="kw">legend.direction</span> <span class="kw">=</span> <span class="st">"vertical"</span>))
<span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/ggdraw.html">ggdraw</a></span>() +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/draw_plot.html">draw_plot</a></span>(<span class="no">a</span>, <span class="kw">x</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">y</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">width</span> <span class="kw">=</span> <span class="fl">0.4</span>, <span class="kw">height</span> <span class="kw">=</span> <span class="fl">1</span>) +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/draw_plot.html">draw_plot</a></span>(<span class="no">b</span> + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="kw">color</span><span class="kw">=</span><span class="no">F</span>, <span class="kw">shape</span><span class="kw">=</span><span class="no">F</span>), <span class="kw">x</span><span class="kw">=</span><span class="fl">0.42</span>, <span class="kw">y</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">width</span> <span class="kw">=</span> <span class="fl">0.4</span>, <span class="kw">height</span> <span class="kw">=</span> <span class="fl">1</span>) +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/draw_grob.html">draw_grob</a></span>(<span class="no">leg</span>, <span class="kw">x</span><span class="kw">=</span><span class="fl">0.4</span>, <span class="kw">y</span><span class="kw">=</span><span class="fl">0</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/beforeafter2-1.png" width="960" style="display: block; margin: auto;"></p>
<p>The resulting figure demonstrates that removing the various contaminants flagged above also increases the dissimilarities of <em>samples</em> both within and between- habitat and material types.</p>
</div>
<div id="data-aggregation" class="section level3">
<h3 class="hasAnchor">
<a href="#data-aggregation" class="anchor"></a>Data aggregation</h3>
<p>At this stage, technical replicates of <em>pcrs</em> should be aggregated so as to focus on biological patterns. This can be done with the function <code>aggregate_pcrs</code>, which includes flexible possibilities for data aggregation methods. Here, we will simply sum MOTUs counts across <em>pcrs</em> replicates, which is the default function of <code>aggregate_pcrs</code>. Note that this function can also be used for other purposes than for aggregating technical replicates. For example, if technical replicates are not included in the experiment, one could instead aggregate biological replicates from a same site.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">soil_euk_agg</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/aggregate_pcrs.html">aggregate_pcrs</a></span>(<span class="no">soil_euk_clean</span>)
<span class="fu"><a href="../reference/summary_metabarlist.html">summary_metabarlist</a></span>(<span class="no">soil_euk_agg</span>)
<span class="co">#&gt; $dataset_dimension</span>
<span class="co">#&gt;         n_row n_col</span>
<span class="co">#&gt; reads      61 10570</span>
<span class="co">#&gt; motus   10570    18</span>
<span class="co">#&gt; pcrs       61    20</span>
<span class="co">#&gt; samples    61     8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $dataset_statistics</span>
<span class="co">#&gt;         nb_reads nb_motus avg_reads sd_reads avg_motus sd_motus</span>
<span class="co">#&gt; pcrs     2193430    10570  35957.87 21718.55  965.0328 401.6088</span>
<span class="co">#&gt; samples  2193430    10570  35957.87 21718.55  965.0328 401.6088</span></pre></body></html></div>
<p>Now, <code>soil_euk_agg</code> has the same dataset characteristics for <em>pcrs</em> and <em>samples</em>.</p>
</div>
</div>
<div id="final-visualisations" class="section level2">
<h2 class="hasAnchor">
<a href="#final-visualisations" class="anchor"></a>Final visualisations</h2>
<p>Our data is now ready for ecological analyses with other <code>R</code> packages. Before finishing this tutorial, we will demonstrate the use of two last <code>metabaR</code> functions in addition to those of other <code>R</code>packages to roughly explore the characteristics of the final dataset.</p>
<div id="taxonomic-composition" class="section level3">
<h3 class="hasAnchor">
<a href="#taxonomic-composition" class="anchor"></a>Taxonomic composition</h3>
<p>One challenge when assessing the taxonomic composition of DNA metabarcoding data is (i) taxonomic information that is not always available at the same taxonomic resolution, and (ii) the large taxonomic breadth of the community under study. The <code>ggtaxplot</code> function builds a taxonomic tree from information contained in the <code>motus</code> table, in order to provide an overview of the composition of the whole community:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="../reference/ggtaxplot.html">ggtaxplot</a></span>(<span class="no">soil_euk_agg</span>,
          <span class="kw">taxo</span> <span class="kw">=</span> <span class="st">"path"</span>,
          <span class="kw">sep.level</span> <span class="kw">=</span> <span class="st">":"</span>, <span class="kw">sep.info</span> <span class="kw">=</span> <span class="st">"@"</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/taxo1-1.png" width="960" style="display: block; margin: auto;"></p>
<p>The taxonomic tree above shows that the <code>soil_euk</code> dataset is mainy composed of Fungi and Metazoa MOTUs and reads. Let’s have a further look at the taxonomic composition of phyla in one kingdom, the Metazoa. This can be done with function <code>aggregate_motus</code>. The kingdom level is not available as a single column in the <code>soil_euk</code> dataset, and so requires the parsing of the full MOTU taxonomic information available in the column “path”, with the <code>taxoparser</code> function.</p>
<div class="sourceCode"><html><body><pre class="r"><span class="co"># Get the kingdom names of MOTUs</span>
<span class="no">kingdom</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/unname.html">unname</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fu"><a href="../reference/taxoparser.html">taxoparser</a></span>(<span class="kw">taxopath</span> <span class="kw">=</span> <span class="no">soil_euk_agg</span>$<span class="no">motus</span>$<span class="no">path</span>,
                           <span class="kw">sep.level</span> <span class="kw">=</span> <span class="st">":"</span>,<span class="kw">sep.info</span> <span class="kw">=</span> <span class="st">"@"</span>),
                <span class="kw">function</span>(<span class="no">x</span>) <span class="no">x</span>[<span class="st">"kingdom"</span>]))

<span class="co"># Create metazoa metabarlist</span>
<span class="no">soil_metazoa</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/subset_metabarlist.html">subset_metabarlist</a></span>(<span class="no">soil_euk_agg</span>, <span class="st">"motus"</span>,
                                    <span class="no">kingdom</span><span class="kw">==</span><span class="st">"Metazoa"</span> <span class="kw">&amp;</span> !<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">kingdom</span>))

<span class="co"># Aggregate the metabarlist at the phylum level</span>
<span class="no">soil_metazoa_phy</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="../reference/aggregate_motus.html">aggregate_motus</a></span>(<span class="no">soil_metazoa</span>, <span class="kw">groups</span> <span class="kw">=</span> <span class="no">soil_metazoa</span>$<span class="no">motus</span>$<span class="no">phylum_name</span>)

<span class="co"># Aggregate per Habitat/material</span>
<span class="no">tmp</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span>(<span class="no">soil_metazoa_phy</span>$<span class="no">reads</span>, <span class="kw">by</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="no">soil_metazoa_phy</span>$<span class="no">samples</span>$<span class="no">Habitat</span>,
          <span class="no">soil_metazoa_phy</span>$<span class="no">samples</span>$<span class="no">Material</span>,
          <span class="kw">sep</span> <span class="kw">=</span> <span class="st">" | "</span>)), <span class="no">sum</span>))

<span class="co"># Plot </span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">tmp</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">Group.1</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">value</span>, <span class="kw">fill</span><span class="kw">=</span><span class="no">variable</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>(<span class="kw">stat</span><span class="kw">=</span><span class="st">"identity"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="kw">NULL</span>, <span class="kw">y</span><span class="kw">=</span><span class="st">"#reads"</span>, <span class="kw">fill</span><span class="kw">=</span><span class="st">"Phyla"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="kw">palette</span> <span class="kw">=</span> <span class="st">"Set3"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="kw">legend.position</span> <span class="kw">=</span> <span class="st">"bottom"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Metazoa phyla"</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/taxo2-1.png" width="768" style="display: block; margin: auto;"></p>
<p>By considering a subset of our dataset, and aggregating at specific taxonomic levels, this approach has allowed us to uncover some initial trends in our data (e.g. that arthropod reads are higher in our soil samples versus our leaf litter samples).</p>
</div>
<div id="rarefaction-curves" class="section level3">
<h3 class="hasAnchor">
<a href="#rarefaction-curves" class="anchor"></a>Rarefaction curves</h3>
<p>To finish, lets take a final look at the diversity coverage of <em>samples</em> with rarefaction curves (rather than our previous look at the <em>pcrs</em> level).</p>
<div class="sourceCode"><html><body><pre class="r"><span class="no">soil_euk_agg.raref</span> <span class="kw">=</span> <span class="fu"><a href="../reference/hill_rarefaction.html">hill_rarefaction</a></span>(<span class="no">soil_euk_agg</span>, <span class="kw">nboot</span> <span class="kw">=</span> <span class="fl">20</span>, <span class="kw">nsteps</span> <span class="kw">=</span> <span class="fl">10</span>)
<span class="no">material</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="no">soil_euk_agg</span>$<span class="no">samples</span>$<span class="no">Habitat</span>, <span class="no">soil_euk_agg</span>$<span class="no">samples</span>$<span class="no">Material</span>)
<span class="no">material</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="no">material</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">soil_euk_agg</span>$<span class="no">samples</span>))

<span class="co">#plot</span>
<span class="no">p</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/hill_rarefaction.html">gghill_rarefaction</a></span>(<span class="no">soil_euk_agg.raref</span>, <span class="kw">group</span><span class="kw">=</span><span class="no">material</span>)
<span class="no">p</span> + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"brown4"</span>, <span class="st">"brown1"</span>, <span class="st">"goldenrod4"</span>, <span class="st">"goldenrod1"</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"brown4"</span>, <span class="st">"brown1"</span>, <span class="st">"goldenrod4"</span>, <span class="st">"goldenrod1"</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">color</span><span class="kw">=</span><span class="st">"Habitat | Material type"</span>)</pre></body></html></div>
<p><img src="metabaRF-vignette_files/figure-html/raref-1.png" width="768" style="display: block; margin: auto;"></p>
<p>This visualisation allows us to explore patterns of diversity across our sample substrates and habitats. You can now go on with ecological analyses using <code>metabaR</code> for data handling and any other R package!</p>
<p>Hopefully this tutorial has demonstrated the strength of <code>metabaR</code> in visualising data, its use for identifying and removing surious sequences, and ultimately its complementary nature with existing packages for data handling. We hope that you are encouraged to incorporate the above ideas into your future processing of metabarcoding data!</p>
</div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-alberdi:2019:00">
<p>Alberdi, Antton, and M Thomas P Gilbert. 2019. “A Guide to the Application of Hill Numbers to Dna-Based Diversity Analyses.” <em>Mol Ecol Resour</em> 19 (4): 804–17. doi:<a href="https://doi.org/10.1111/1755-0998.13014">10.1111/1755-0998.13014</a>.</p>
</div>
<div id="ref-boyer:2016:00">
<p>Boyer, F., C. Mercier, A. Bonin, Y. Le Bras, P. Taberlet, and E. Coissac. 2016. “OBITools : A Unix- Inspired Software Package for Dna Metabarcoding.” <em>Molecular Ecology Resources</em> 16 (1): 176–82.</p>
</div>
<div id="ref-calderon2020environmental">
<p>Calderón-Sanou, Irene, Tamara Münkemüller, Frédéric Boyer, Lucie Zinger, and Wilfried Thuiller. 2020. “From Environmental Dna Sequences to Ecological Conclusions: How Strong Is the Influence of Methodological Choices?” <em>Journal of Biogeography</em> 47. Wiley Online Library: 193–206.</p>
</div>
<div id="ref-chao2014unifying">
<p>Chao, Anne, Chun-Huo Chiu, and Lou Jost. 2014. “Unifying Species Diversity, Phylogenetic Diversity, Functional Diversity, and Related Similarity and Differentiation Measures Through Hill Numbers.” <em>Annual Review of Ecology, Evolution, and Systematics</em> 45. Annual Reviews: 297–324.</p>
</div>
<div id="ref-esling:2015:00">
<p>Esling, Philippe, Franck Lejzerowicz, and Jan Pawlowski. 2015. “Accurate Multiplexing and Filtering for High-Throughput Amplicon-Sequencing.” <em>Nucleic Acids Research</em> 43 (5): 2513–24. doi:<a href="https://doi.org/10.1093/nar/gkv107">10.1093/nar/gkv107</a>.</p>
</div>
<div id="ref-mercier:2013:00">
<p>Mercier, C., F. Boyer, E. Kopylova, P. Taberlet, A. Bonin, and E. Coissac. 2013. “SUMATRA and Sumaclust: Fast and Exact Comparison and Clustering of Sequences.” <em>Programs and Abstracts of the SeqBio 2013 Workshop</em>, 27–29.</p>
</div>
<div id="ref-quast2013silva">
<p>Quast, Christian, Elmar Pruesse, Pelin Yilmaz, Jan Gerken, Timmy Schweer, Pablo Yarza, Jörg Peplies, and Frank Oliver Glöckner. 2013. “The Silva Ribosomal Rna Gene Database Project: Improved Data Processing and Web-Based Tools.” <em>Nucleic Acids Research</em> 41 (D1). Oxford University Press: D590–D596.</p>
</div>
<div id="ref-schnell:2015:00">
<p>Schnell, Ida Bærholm, Kristine Bohmann, and M. Thomas P. Gilbert. 2015. “Tag Jumps Illuminated - Reducing Sequence-to-Sample Misidentifications in Metabarcoding Studies.” <em>Molecular Ecology Resources</em>, n/a–n/a. doi:<a href="https://doi.org/10.1111/1755-0998.12402">10.1111/1755-0998.12402</a>.</p>
</div>
<div id="ref-taberlet2018environmental">
<p>Taberlet, Pierre, Aurélie Bonin, Lucie Zinger, and Eric Coissac. 2018. <em>Environmental Dna: For Biodiversity Research and Monitoring</em>. Oxford University Press.</p>
</div>
<div id="ref-zinger2019dna">
<p>Zinger, Lucie, Aurélie Bonin, Inger G Alsos, Miklós Bálint, Holly Bik, Frédéric Boyer, Anthony A Chariton, et al. 2019. “DNA Metabarcoding—Need for Robust Experimental Designs to Draw Sound Ecological Conclusions.” <em>Molecular Ecology</em> 28 (8). Wiley Online Library: 1857–62.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Lucie Zinger, Frédéric Boyer, Clément Lionnet, Anne-Sophie Benoiston.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
