<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>The karst dataset: sampler effect and technical reproducibility • metabaR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="The karst dataset: sampler effect and technical reproducibility">
<meta property="og:description" content="metabaR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metabaR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/metabaRF-vignette.html">Let's metabaR!</a>
    </li>
    <li>
      <a href="../articles/metabaRF-vignette_karst.html">The karst dataset: sampler effect and technical reproducibility</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/metabaRfactory/metabaR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="metabaRF-vignette_karst_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="metabaRF-vignette_karst_files/anchor-sections-1.0/anchor-sections.js"></script><script src="metabaRF-vignette_karst_files/kePrint-0.0.1/kePrint.js"></script><link href="metabaRF-vignette_karst_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>The karst dataset: sampler effect and technical reproducibility</h1>
            
            <h4 class="date">2020-12-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/metabaRfactory/metabaR/blob/master/vignettes/metabaRF-vignette_karst.Rmd"><code>vignettes/metabaRF-vignette_karst.Rmd</code></a></small>
      <div class="hidden name"><code>metabaRF-vignette_karst.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>What if the sampler had an effect on DNA metabarcoding results? What is the extent of technical variability in DNA metabarcoding experiments? To further illustrate how one can further use <code>metabaR</code> to manipulate DNA metabarcoding data, we will address these questions with an new dataset, available in the <code>metabaR</code> companion data repository: <a href="https://github.com/metabaRfactory/metabaR_external_data" class="uri">https://github.com/metabaRfactory/metabaR_external_data</a>.</p>
</div>
<div id="the-karst_euk-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#the-karst_euk-dataset" class="anchor"></a>The <code>karst_euk</code> dataset</h1>
<p>The <code>karst_euk</code> dataset is an object of class <code>metabarlist</code>. The data were obtained from an environmental DNA (eDNA) metabarcoding experiment aiming to assess the turnover of soil eukaryotes in karsts ecosystems, often seen as “terrestrial islands”. In these ecosystems, dispersal is stronly limited by the singular topology of the area and genetic/ecological drifts are expected to be important processes locally.</p>
<p>A sampling was conducted across several kast mountains in Laos, as shown in the figure below. At each sampling site, two composite soil samples were made by two different samplers, one by Jean and one by Yvette. These composite samples are a mix of soil cores collected along a transect, as indicated in the figure. Each of these composite soils were then subjected to DNA extraction in duplicates, using an extracellular DNA extraction protocol <span class="citation">(Taberlet et al. 2018)</span>. On each DNA extact, the v7 region of the 18S rRNA gene was then amplified by PCR in duplicates following previously described protocols <span class="citation">(Zinger et al. 2019)</span> and the Euka02 primer pair <span class="citation">(Taberlet et al. 2018)</span>, and finally sequenced on an HiSeq Illumina platform with the paired-end technology. The retrieved data were then processed using the OBITools <span class="citation">(Boyer et al. 2016)</span> package, including denoising using the <code>obiclean</code> command. Each entry in this table is the equivalent of an ASV (amplicon sequence variant) or the representative sequence of a MOTU.</p>
<p><img src="karst_euk.jpeg" style="width:100.0%"></p>
<p>Given this set up, one expect strong differences in community composition across sites, caused by dispersal limitation, and - hopefully - minimal technical variation caused either by the extraction, PCR, and sampler.</p>
</div>
<div id="data-import" class="section level1">
<h1 class="hasAnchor">
<a href="#data-import" class="anchor"></a>Data import</h1>
<p>The dataset can be imported in <code>R</code> as follows</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/metabaRfactory/metabaR_external_data/master/"</span>
<span class="va">karst_euk_file</span> <span class="op">&lt;-</span> <span class="st">"karst_euk.rds"</span>
<span class="va">karst_euk_url</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">karst_euk_file</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span>
<span class="va">karst_euk_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">karst_euk_file</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">karst_euk_url</span>, <span class="va">karst_euk_path</span><span class="op">)</span>

<span class="va">karst_euk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">karst_euk_path</span><span class="op">)</span></code></pre></div>
<p>One can then use <code>metabaR</code> to display the summary statistics of the <code>karst_euk</code> dataset</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/metabaRfactory/metabaR">metabaR</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/summary_metabarlist.html">summary_metabarlist</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">)</span>
<span class="co">#&gt; $dataset_dimension</span>
<span class="co">#&gt;         n_row n_col</span>
<span class="co">#&gt; reads      48  7870</span>
<span class="co">#&gt; motus    7870    23</span>
<span class="co">#&gt; pcrs       48     4</span>
<span class="co">#&gt; samples    12     5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $dataset_statistics</span>
<span class="co">#&gt;         nb_reads nb_motus avg_reads sd_reads avg_motus sd_motus</span>
<span class="co">#&gt; pcrs      224627     7870  4679.729 1340.161  933.1667 312.2533</span>
<span class="co">#&gt; samples   224627     7870  4679.729 1340.161  933.1667 312.2533</span></code></pre></div>
<p>The <code>karst_euk</code> dataset consists of 48 PCR products corresponding to 12 composite, biological samples in total. It does not contains experimental negative or positive controls, which is why the summary statistics are the same between <em>pcrs</em> and <em>samples</em>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">)</span>
<span class="co">#&gt; [1] "sample_id"     "extraction_id" "type"          "control_type"</span></code></pre></div>
<p>In karst_euk, the columns of the <code>pcrs</code> table correspond to:</p>
<ul>
<li>
<code>"sample_id"</code>: a vector indicating the biological sample origin of each <em>pcr</em> (e.g. the sample name), here a composite sample collected by a given sampler.<br>
</li>
<li>
<code>"extraction_id"</code>: a vector indicating the extraction origin of each <em>pcr</em> (duplicate extraction for each <em>sample</em>).<br>
</li>
<li>
<code>"type"</code>: a vector containing only the value “sample”, as no controls are included in this experiment.<br>
</li>
<li>
<code>"control_type"</code>: a vector containing only the value <code>NA</code>, as no controls are included in this experiment.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span>
<span class="co">#&gt; [1] "Site"    "avg_N"   "avg_E"   "avg_alt" "sampler"</span></code></pre></div>
<p>The columns of the <code>samples</code> correspond to: - <code>"Site"</code>: the sampling site.<br>
- <code>"avg_N"</code>: the latitude (decimal degrees) averaged across each sampling points of the same site.<br>
- <code>"avg_E"</code>: the longitude (decimal degrees), averaged as for latitudes. - <code>"avg_alt"</code>: the elevation (m.a.s.l.), averaged as above.<br>
- <code>"sampler"</code>: the name of the sampler.</p>
</div>
<div id="diagnostic-plots" class="section level1">
<h1 class="hasAnchor">
<a href="#diagnostic-plots" class="anchor"></a>Diagnostic plots</h1>
<div id="basic-visualisation" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-visualisation" class="anchor"></a>Basic visualisation</h2>
<p>As the dataset does not contain neither negative / positive controls, nor information on the pcr design (i.e. PCR plates and wells), several functions of the <code>metabaR</code> package cannot be used. Still, it enables handling this multi-layer data for more basic visualisation purposes. For example, the distribution of sequencing depth and number of MOTUs per samples can be visualized as follows</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Compute the number of reads per pcr</span>
<span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">nb_reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">reads</span><span class="op">)</span>

<span class="co"># Compute the number of motus per pcr</span>
<span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">nb_motus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">reads</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span>

<span class="co"># load ggplot2</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>

<span class="co"># Plot the distribution of #reads </span>
<span class="va">a</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">nb_reads</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"grey"</span>, fill<span class="op">=</span><span class="st">"white"</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expand_limits.html">expand_limits</a></span><span class="op">(</span>x<span class="op">=</span><span class="fl">0</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"# reads"</span>, y<span class="op">=</span><span class="st">"# PCRs"</span><span class="op">)</span>

<span class="co"># Plot the distribution of #motus</span>
<span class="va">b</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">nb_motus</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"grey"</span>, fill<span class="op">=</span><span class="st">"white"</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"# MOTUs"</span>, y<span class="op">=</span><span class="st">"# PCRs"</span><span class="op">)</span>

<span class="co"># Combine plots into one</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span>
<span class="fu"><a href="https://wilkelab.org/cowplot/reference/ggdraw.html">ggdraw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html">draw_plot</a></span><span class="op">(</span><span class="va">a</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0</span>, width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html">draw_plot</a></span><span class="op">(</span><span class="va">b</span>, x<span class="op">=</span><span class="fl">0.5</span>, y<span class="op">=</span><span class="fl">0</span>, width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/diag1-1.png" width="576" style="display: block; margin: auto;"></p>
<p>This very basics plots already suggest that 3 <em>pcrs</em> have a very low diversity. They somehow behave like negative controls where some contaminants have been amplified.</p>
</div>
<div id="sequencing-depth-and-coverage" class="section level2">
<h2 class="hasAnchor">
<a href="#sequencing-depth-and-coverage" class="anchor"></a>Sequencing depth and coverage</h2>
<p>Let’s construct the rarefaction curves with the <code>hill_rarefaction</code> function so that to obtain an estimate of <span class="math inline">\(^{q}D\)</span> and <span class="math inline">\(Coverage\)</span>. <code>nboot</code> is low in the example below to limit the computing time.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#build rarefaction curves</span>
<span class="va">karst_euk.raref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hill_rarefaction.html">hill_rarefaction</a></span><span class="op">(</span><span class="va">karst_euk</span>, nboot <span class="op">=</span> <span class="fl">20</span>, nsteps <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>


<span class="co">#draw by coloring the curves by sampler</span>
<span class="co"># Define a vector containing the Material info for each pcrs </span>
<span class="va">sampler</span> <span class="op">&lt;-</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">sampler</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sample_id</span>,
                                                <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>

<span class="co"># Use of gghill_rarefaction requires a vector with named pcrs</span>
<span class="va">sampler</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span><span class="va">sampler</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Plot</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hill_rarefaction.html">gghill_rarefaction</a></span><span class="op">(</span><span class="va">karst_euk.raref</span>, group<span class="op">=</span><span class="va">sampler</span><span class="op">)</span>
<span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"orange"</span>,<span class="st">"chartreuse"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"orange"</span>,<span class="st">"chartreuse"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"Sampler"</span><span class="op">)</span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/hill-1.png" width="672" style="display: block; margin: auto;"></p>
<p>At a first glance, it seems that Jean’s samples are more diverse in abundant/intermediate sequences than Yvette’s ones.</p>
</div>
</div>
<div id="flagging-spurious-motus" class="section level1">
<h1 class="hasAnchor">
<a href="#flagging-spurious-motus" class="anchor"></a>Flagging spurious MOTUs</h1>
<div id="potential-contaminants" class="section level2">
<h2 class="hasAnchor">
<a href="#potential-contaminants" class="anchor"></a>Potential contaminants</h2>
<p>This dataset does not include negative controls to enable detecting potential contaminants objectively. Still, we identified above some <em>pcrs</em> with very low diversity that may correspond to <em>pcrs</em> without DNA template and where reagents contaminants were amplified. Let’s see if these have a particular composition as compared to the other samples. This can be done by hijacking the <code>contaslayer</code> function as follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># set low-diversity pcrs as controls of e.g. pcrs</span>
<span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">control_type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">nb_motus</span><span class="op">&lt;</span><span class="fl">100</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"pcr"</span>
<span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">nb_motus</span><span class="op">&lt;</span><span class="fl">100</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"control"</span>

<span class="co"># run contaslayer</span>
<span class="va">karst_euk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/contaslayer.html">contaslayer</a></span><span class="op">(</span><span class="va">karst_euk</span>, 
                         control_types <span class="op">=</span> <span class="st">"pcr"</span>,
                         output_col <span class="op">=</span> <span class="st">"not_a_pcr_conta"</span><span class="op">)</span></code></pre></div>
<p>And then display some basic information on these potential contaminating sequences.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># print taxonomy of MOTUs most abundant in low-diversity samples</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/">kableExtra</a></span><span class="op">)</span>
<span class="va">dt</span> <span class="op">&lt;-</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">[</span><span class="op">!</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_a_pcr_conta</span>,
                     <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"count"</span>,<span class="st">"best_identity"</span>,
                       <span class="st">"scientific_name"</span>, <span class="st">"sequence"</span><span class="op">)</span><span class="op">]</span>
<span class="va">dt</span><span class="op">$</span><span class="va">best_identity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">dt</span><span class="op">$</span><span class="va">best_identity</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">dt</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"total # reads"</span>, <span class="st">"similarity to ref DB"</span>, <span class="st">"taxon name"</span>, <span class="st">"sequence"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">dt</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">dt</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,<span class="op">]</span>, row.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html">kable_styling</a></span><span class="op">(</span>bootstrap_options<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"striped"</span>, <span class="st">"hover"</span>, <span class="st">"condensed"</span><span class="op">)</span>, 
                font_size <span class="op">=</span> <span class="fl">8</span>, full_width <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre></div>
<table class="table table table-striped table-hover table-condensed">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
total # reads
</th>
<th style="text-align:right;">
similarity to ref DB
</th>
<th style="text-align:left;">
taxon name
</th>
<th style="text-align:left;">
sequence
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
laos_18Seuka_00044
</td>
<td style="text-align:right;">
2752
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Sordariomycetes
</td>
<td style="text-align:left;">
ctcaaacttccatccgcttgagcggatagtccctctaagaagccagcgtactgccaaagcaatacgggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00076
</td>
<td style="text-align:right;">
1768
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Liliopsida
</td>
<td style="text-align:left;">
ctcaaacttccgtggcctaaacggccatagtccctctaagaagctagctgcggagggatggctccgcatagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00048
</td>
<td style="text-align:right;">
738
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Pooideae
</td>
<td style="text-align:left;">
ctcaaacttccgtcgcctaaacggcgatagtccctctaagaagctagctgcggagggatggctccgcatagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00390
</td>
<td style="text-align:right;">
437
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Euclea crispa subsp. linearis
</td>
<td style="text-align:left;">
ctcaaacttccgtggcctgaaaggccatagtccctctaagaagctagctgcggagggtcgcctccgcatagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00379
</td>
<td style="text-align:right;">
400
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Brassicaceae
</td>
<td style="text-align:left;">
ctcaaacttccttggcctaaacggccatagtccctctaagaagccggccgtgaagggatgcctccacgtagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00098
</td>
<td style="text-align:right;">
266
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Euplotida
</td>
<td style="text-align:left;">
ctcaaacttccttgtggttgcacacaaagtccctctaagaagtacataccggaaaaccggctgactatttagcaggctaaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00366
</td>
<td style="text-align:right;">
248
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Basidiomycota
</td>
<td style="text-align:left;">
ctcaaacttccgtcagctaaacgctgacagtccctctaagaagccagcgaccagcaaaagccggccgggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00329
</td>
<td style="text-align:right;">
220
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Diplosoma gumavirens
</td>
<td style="text-align:left;">
ctcaaacttccattggctggaagccaatagtccctctaagaagccagccaccaaccatagtcgatggggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00866
</td>
<td style="text-align:right;">
176
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Fungi
</td>
<td style="text-align:left;">
ctcaaacttccttcggcttgagccgaaagtccctctaagaagccagcgtactgccaaagcaatacgggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_01002
</td>
<td style="text-align:right;">
174
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Vorticella campanula
</td>
<td style="text-align:left;">
ctcaaacttccatgtgattacatcacatagtccctctaagaagtgattcaaatttgaaataagaacactagttagcaggttaaggtctcgttcgttaa
</td>
</tr>
</tbody>
</table>
<p>From the taxonomy of these potential contaminant sequences, it is difficult to say whether these sequences are actual contaminants. One way to further investigate this is to see how these sequences are distributed across <em>pcrs</em>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape">reshape2</a></span><span class="op">)</span>
<span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">reads</span><span class="op">[</span>,<span class="op">!</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_a_pcr_conta</span><span class="op">]</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">tmp</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Var2</span>, y<span class="op">=</span><span class="va">Var1</span>, size<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">value</span><span class="op">==</span><span class="fl">0</span>, <span class="cn">NA</span>, <span class="va">value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>size <span class="op">=</span> <span class="st">"# reads"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/conta3-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The most abundant contaminant identified in the table above is present in all samples, and is therefore most likely a contaminant indeed. In addition other potential contaminant sequences are present in 7-8 successive samples, which are likely in the same column of a PCR plate. These features strongly suggest that these sequences are indeed artefactual and should be excluded from the dataset.</p>
</div>
<div id="flagging-spurious-motus-and-non-target-motus" class="section level2">
<h2 class="hasAnchor">
<a href="#flagging-spurious-motus-and-non-target-motus" class="anchor"></a>Flagging spurious MOTUs and non-target MOTUs</h2>
<p>Non-target MOTUs cannot be detected in this particular dataset, because the taxonomic assignments have been made using a reference database that only contains eukaryotic sequences.</p>
<p>One can, however, still identify MOTUs whose sequence is too dissimilar from references, by drawing the distribution of MOTU similarity scores, weighted and unweighted by their relative abundance.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot the unweighted distribution of MOTUs similarity scores </span>
<span class="va">a</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">best_identity</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"grey"</span>, fill<span class="op">=</span><span class="st">"white"</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.8</span>, col<span class="op">=</span><span class="st">"orange"</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"% similarity against best match"</span>, y<span class="op">=</span><span class="st">"# MOTUs"</span><span class="op">)</span>

<span class="co"># Same for the weighted distribution</span>
<span class="va">b</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span>, 
         <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">best_identity</span>, y <span class="op">=</span> <span class="va">..count..</span>, weight <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"grey"</span>, fill<span class="op">=</span><span class="st">"white"</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.8</span>, col<span class="op">=</span><span class="st">"orange"</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"% similarity against best match"</span>, y<span class="op">=</span><span class="st">"# Reads"</span><span class="op">)</span>

<span class="co"># Combine plots into one</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span>
<span class="fu"><a href="https://wilkelab.org/cowplot/reference/ggdraw.html">ggdraw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html">draw_plot</a></span><span class="op">(</span><span class="va">a</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0</span>, width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html">draw_plot</a></span><span class="op">(</span><span class="va">b</span>, x<span class="op">=</span><span class="fl">0.5</span>, y<span class="op">=</span><span class="fl">0</span>, width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/nontarget3-1.png" width="576" style="display: block; margin: auto;"></p>
<p>As in the <code>soil_euk</code> dataset, we may consider any MOTU as degraded sequences if its sequence similarity is &lt; 80% against its best match in the reference database.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Flag not degraded (TRUE) vs. potentially degraded sequences (FALSE)</span>
<span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_degraded</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">best_identity</span> <span class="op">&lt;</span> <span class="fl">0.8</span>, <span class="cn">F</span>, <span class="cn">T</span><span class="op">)</span>

<span class="co"># Proportion of each of these over total number of MOTUs</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_degraded</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;      FALSE       TRUE </span>
<span class="co">#&gt; 0.02210928 0.97789072</span>

<span class="co"># Intersection with other flags</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_a_pcr_conta</span>, 
      <span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_degraded</span><span class="op">)</span>
<span class="co">#&gt;        </span>
<span class="co">#&gt;         FALSE TRUE</span>
<span class="co">#&gt;   FALSE     0   40</span>
<span class="co">#&gt;   TRUE    174 7656</span></code></pre></div>
<p>No intersection is found between degraded MOTUs and potential contaminants.</p>
</div>
</div>
<div id="assessing-signal-reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#assessing-signal-reproducibility" class="anchor"></a>Assessing signal reproducibility</h1>
<p>Here, one can assess the signal reproducibility at different levels:<br>
- the DNA extract level<br>
- the composite sample level<br>
- the site level.</p>
<p>Because the functions in <code>metabaR</code> rely on</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># DNA extraction</span>
<span class="va">comp1</span> <span class="op">=</span> <span class="fu"><a href="../reference/pcrslayer.html">pcr_within_between</a></span><span class="op">(</span><span class="va">karst_euk</span>, replicates <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">extraction_id</span><span class="op">)</span>
<span class="va">a</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_thresh</a></span><span class="op">(</span><span class="va">comp1</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"between DNA extracts"</span>, <span class="st">"within DNA extracts"</span><span class="op">)</span>, 
                       values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"coral2"</span>, <span class="st">"cyan3"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># composite sample</span>
<span class="va">comp2</span> <span class="op">=</span> <span class="fu"><a href="../reference/pcrslayer.html">pcr_within_between</a></span><span class="op">(</span><span class="va">karst_euk</span>, replicates <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span>
<span class="va">b</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_thresh</a></span><span class="op">(</span><span class="va">comp2</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"between samples"</span>, <span class="st">"within DNA samples"</span><span class="op">)</span>, 
                       values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"coral2"</span>, <span class="st">"cyan3"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># sampler</span>
<span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sampler_id</span> <span class="op">&lt;-</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">sampler</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sample_id</span>, 
                                                             <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="va">comp3</span> <span class="op">=</span> <span class="fu"><a href="../reference/pcrslayer.html">pcr_within_between</a></span><span class="op">(</span><span class="va">karst_euk</span>, replicates <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sampler_id</span><span class="op">)</span>

<span class="va">comp3</span><span class="op">$</span><span class="va">bar_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="va">comp3</span><span class="op">$</span><span class="va">bar_dist</span>, <span class="fl">2</span><span class="op">)</span> <span class="co"># trick to plot a density from 1 value</span>

<span class="va">c</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_thresh</a></span><span class="op">(</span><span class="va">comp3</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"between samplers"</span>, <span class="st">"within DNA samplers"</span><span class="op">)</span>, 
                       values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"coral2"</span>, <span class="st">"cyan3"</span><span class="op">)</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span>
<span class="fu"><a href="https://wilkelab.org/cowplot/reference/ggdraw.html">ggdraw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html">draw_plot</a></span><span class="op">(</span><span class="va">a</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0.6</span>, height <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html">draw_plot</a></span><span class="op">(</span><span class="va">b</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0.3</span>,  height <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html">draw_plot</a></span><span class="op">(</span><span class="va">c</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0</span>,  height <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/pcrsalyer1-1.png" width="480" style="display: block; margin: auto;"></p>
<p>These plots suggest that the PCR and extraction reproducibility is high. By contrast, reproducibility per sample is very low, which means that biological difference between samples are stronger than a potential effect introduced by the sampler per se.</p>
<p>Let’s verify this with an ordination visualisation by hijacking the <code>check_pcr_repl</code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Distinguish between pcrs obtained from the different sampling sites</span>
<span class="va">mds</span> <span class="op">=</span> <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_repl</a></span><span class="op">(</span><span class="va">karst_euk</span>, 
                     <span class="co"># display sites with different colors</span>
                     groups <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Site</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sample_id</span>,
                                                           <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>, 
                     <span class="co"># use funcpcr to differentiate samplers instead of good vs. failed pcrs.</span>
                     funcpcr <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sampler_id</span> <span class="op">==</span> <span class="st">"Yvette"</span><span class="op">)</span> 

<span class="va">mds</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Sites"</span>, shape<span class="op">=</span><span class="st">"Sampler"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_shape.html">scale_shape</a></span><span class="op">(</span>labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Jean"</span>, <span class="st">"Yvette"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/pcrsalyer3-1.png" width="480" style="display: block; margin: auto;"></p>
<p>The variability across sites is hence far larger than between samplers and technical replicates. But the sampler effect is not totally negligible.</p>
</div>
<div id="data-final-cleaning-and-aggregation" class="section level1">
<h1 class="hasAnchor">
<a href="#data-final-cleaning-and-aggregation" class="anchor"></a>Data final cleaning and aggregation</h1>
<p>We refer to the <code>soil_euk</code> tutorial for these last steps.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-boyer:2016:00">
<p>Boyer, F., C. Mercier, A. Bonin, Y. Le Bras, P. Taberlet, and E. Coissac. 2016. “OBITools : A Unix- Inspired Software Package for Dna Metabarcoding.” <em>Molecular Ecology Resources</em> 16 (1): 176–82.</p>
</div>
<div id="ref-taberlet2018environmental">
<p>Taberlet, Pierre, Aurélie Bonin, Lucie Zinger, and Eric Coissac. 2018. <em>Environmental Dna: For Biodiversity Research and Monitoring</em>. Oxford University Press.</p>
</div>
<div id="ref-zinger2019body">
<p>Zinger, Lucie, Pierre Taberlet, Heidy Schimann, Aurélie Bonin, Frédéric Boyer, Marta De Barba, Philippe Gaucher, et al. 2019. “Body Size Determines Soil Community Assembly in a Tropical Forest.” <em>Molecular Ecology</em> 28 (3). Wiley Online Library: 528–43.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Lucie Zinger, Frédéric Boyer, Clément Lionnet, Anne-Sophie Benoiston.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
