<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>The karst_euk dataset: sampler effect and technical reproducibility • metabaR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="The karst_euk dataset: sampler effect and technical reproducibility">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metabaR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/metabaRF-vignette_karst.html">The karst_euk dataset: sampler effect and technical reproducibility</a>
    </li>
    <li>
      <a href="../articles/metabaRF-vignette.html">Let's metabaR!</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/metabaRfactory/metabaR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="metabaRF-vignette_karst_files/kePrint-0.0.1/kePrint.js"></script><link href="metabaRF-vignette_karst_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>The karst_euk dataset: sampler effect and
technical reproducibility</h1>
            
            <h4 data-toc-skip class="date">2025-07-31</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/metabaRfactory/metabaR/blob/master/vignettes/metabaRF-vignette_karst.Rmd" class="external-link"><code>vignettes/metabaRF-vignette_karst.Rmd</code></a></small>
      <div class="hidden name"><code>metabaRF-vignette_karst.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>What if the individual collecting the samples affects DNA
metabarcoding results? What is the extent of technical variability in
DNA metabarcoding experiments? To further illustrate how one can use
<code>metabaR</code> to manipulate DNA metabarcoding data, we will
address these questions with a new dataset, available in the
<code>metabaR</code> companion data repository: <a href="https://github.com/metabaRfactory/metabaR_external_data" class="external-link uri">https://github.com/metabaRfactory/metabaR_external_data</a>.</p>
</div>
<div class="section level2">
<h2 id="the-karst_euk-dataset">The <code>karst_euk</code> dataset<a class="anchor" aria-label="anchor" href="#the-karst_euk-dataset"></a>
</h2>
<p>The <code>karst_euk</code> dataset is an object of class
<code>metabarlist</code>. The data were obtained from an environmental
DNA (eDNA) metabarcoding experiment aiming to assess the turnover of
soil eukaryotes in karsts ecosystems, often referred to as “terrestrial
islands”. In these ecosystems, dispersal is strongly limited by the
unique topology of the area, and genetic/ecological drifts are expected
to be important processes locally.</p>
<p>Sampling was conducted across several kast mountains in Laos, as
shown in the figure below. At each sample site, two composite soil
samples were taken by two different samplers, one by Jean and one by
Yvette. These composite samples are a mix of soil cores collected along
a transect, as indicated in the figure. Each of these composite soils
was then subjected to DNA extraction in duplicate, using an
extracellular DNA extraction protocol <span class="citation">(Taberlet
et al. 2018)</span>. On each DNA extract, the v7 region of the 18S rRNA
gene was then amplified by PCR in duplicates following previously
described protocols <span class="citation">(Zinger et al. 2019)</span>,
using the Euka02 primer pair <span class="citation">(Taberlet et al.
2018)</span>, and sequenced on an HiSeq Illumina platform with the
paired-end technology. The retrieved data were then processed using the
OBITools <span class="citation">(Boyer et al. 2016)</span> package,
including denoising using the <code>obiclean</code> command. Each entry
in this table is the equivalent of an ASV (amplicon sequence variant) or
the representative sequence of a MOTU.</p>
<p><img src="karst_euk.jpeg" style="width:100.0%"></p>
<p>Given this setup, one expects strong differences in community
composition across sites, caused by dispersal limitation, and -
hopefully - minimal technical variation caused either by the extraction,
PCR, and sampler.</p>
</div>
<div class="section level2">
<h2 id="data-import">Data import<a class="anchor" aria-label="anchor" href="#data-import"></a>
</h2>
<p>The dataset can be imported in <code>R</code> as follows</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/metabaRfactory/metabaR_external_data/master/"</span></span>
<span><span class="va">karst_euk_file</span> <span class="op">&lt;-</span> <span class="st">"karst_euk.rds"</span></span>
<span><span class="va">karst_euk_url</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">karst_euk_file</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span><span class="va">karst_euk_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">karst_euk_file</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">karst_euk_url</span>, <span class="va">karst_euk_path</span><span class="op">)</span></span>
<span></span>
<span><span class="va">karst_euk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">karst_euk_path</span><span class="op">)</span></span></code></pre></div>
<p>One can then use <code>metabaR</code> to display the summary
statistics of the <code>karst_euk</code> dataset</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/metabaRfactory/metabaR" class="external-link">metabaR</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/summary_metabarlist.html">summary_metabarlist</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">)</span></span>
<span><span class="co">#&gt; $dataset_dimension</span></span>
<span><span class="co">#&gt;         n_row n_col</span></span>
<span><span class="co">#&gt; reads      48  7870</span></span>
<span><span class="co">#&gt; motus    7870    23</span></span>
<span><span class="co">#&gt; pcrs       48     4</span></span>
<span><span class="co">#&gt; samples    12     5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dataset_statistics</span></span>
<span><span class="co">#&gt;         nb_reads nb_motus avg_reads sd_reads avg_motus sd_motus</span></span>
<span><span class="co">#&gt; pcrs      224627     7870  4679.729 1340.161  933.1667 312.2533</span></span>
<span><span class="co">#&gt; samples   224627     7870  4679.729 1340.161  933.1667 312.2533</span></span></code></pre></div>
<p>The <code>karst_euk</code> dataset consists of 48 PCR products
corresponding to 12 composite, biological samples in total. It does not
contain experimental negative or positive controls, which is why the
summary statistics are the same between <em>pcrs</em> and
<em>samples</em>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "sample_id"     "extraction_id" "type"          "control_type"</span></span></code></pre></div>
<p>In karst_euk, the columns of the <code>pcrs</code> table correspond
to:</p>
<ul>
<li>
<code>"sample_id"</code>: a vector indicating the biological sample
origin of each <em>pcr</em> (e.g. the sample name), here a composite
sample collected by a given sampler.<br>
</li>
<li>
<code>"extraction_id"</code>: a vector indicating the extraction
origin of each <em>pcr</em> (duplicate extraction for each
<em>sample</em>).<br>
</li>
<li>
<code>"type"</code>: a vector containing only the value “sample”, as
no controls are included in this experiment.<br>
</li>
<li>
<code>"control_type"</code>: a vector containing only the value
<code>NA</code>, as no controls are included in this experiment.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Site"    "avg_N"   "avg_E"   "avg_alt" "sampler"</span></span></code></pre></div>
<p>The columns of the <code>samples</code> correspond to: -
<code>"Site"</code>: the sampling site.<br>
- <code>"avg_N"</code>: the latitude (decimal degrees) averaged across
each sampling point of the same site.<br>
- <code>"avg_E"</code>: the longitude (decimal degrees), averaged as for
latitudes. - <code>"avg_alt"</code>: the elevation (m.a.s.l.), averaged
as above.<br>
- <code>"sampler"</code>: the name of the sampler.</p>
</div>
<div class="section level2">
<h2 id="diagnostic-plots">Diagnostic plots<a class="anchor" aria-label="anchor" href="#diagnostic-plots"></a>
</h2>
<div class="section level3">
<h3 id="basic-visualisation">Basic visualisation<a class="anchor" aria-label="anchor" href="#basic-visualisation"></a>
</h3>
<p>As the dataset contains neither negative / positive controls, nor
information on the pcr design (i.e. PCR plates and wells), several
functions of the <code>metabaR</code> package cannot be used.
Nevertheless, <code>metabaR</code> can still be used to handle this
multi-layer data for more basic visualisation purposes. For example, the
distribution of sequencing depth and number of MOTUs per samples can be
visualized as follows</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute the number of reads per pcr</span></span>
<span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">nb_reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">reads</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute the number of motus per pcr</span></span>
<span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">nb_motus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">reads</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load ggplot2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the distribution of #reads </span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">nb_reads</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"grey"</span>, fill<span class="op">=</span><span class="st">"white"</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expand_limits.html" class="external-link">expand_limits</a></span><span class="op">(</span>x<span class="op">=</span><span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"# reads"</span>, y<span class="op">=</span><span class="st">"# PCRs"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the distribution of #motus</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">nb_motus</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"grey"</span>, fill<span class="op">=</span><span class="st">"white"</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"# MOTUs"</span>, y<span class="op">=</span><span class="st">"# PCRs"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine plots into one</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/ggdraw.html" class="external-link">ggdraw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html" class="external-link">draw_plot</a></span><span class="op">(</span><span class="va">a</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0</span>, width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html" class="external-link">draw_plot</a></span><span class="op">(</span><span class="va">b</span>, x<span class="op">=</span><span class="fl">0.5</span>, y<span class="op">=</span><span class="fl">0</span>, width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/diag1-1.png" width="576" style="display: block; margin: auto;"></p>
<p>These very basics plots already suggest that 3 <em>pcrs</em> have
very low diversity. They appear to behave in the same way as negative
controls, potentially amplifying only contaminants and few sample site
derived sequences.</p>
</div>
<div class="section level3">
<h3 id="sequencing-depth-and-coverage">Sequencing depth and coverage<a class="anchor" aria-label="anchor" href="#sequencing-depth-and-coverage"></a>
</h3>
<p>Let’s construct the rarefaction curves with the
<code>hill_rarefaction</code> function so as to obtain an estimate of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi></mi><mi>q</mi></msup><mi>D</mi></mrow><annotation encoding="application/x-tex">^{q}D</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>o</mi><mi>v</mi><mi>e</mi><mi>r</mi><mi>a</mi><mi>g</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Coverage</annotation></semantics></math>.
<code>nboot</code> is low in the example below to limit the computing
time.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#build rarefaction curves</span></span>
<span><span class="va">karst_euk.raref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hill_rarefaction.html">hill_rarefaction</a></span><span class="op">(</span><span class="va">karst_euk</span>, nboot <span class="op">=</span> <span class="fl">20</span>, nsteps <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#draw by coloring the curves by sampler</span></span>
<span><span class="co"># Define a vector containing the Material info for each pcrs </span></span>
<span><span class="va">sampler</span> <span class="op">&lt;-</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">sampler</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sample_id</span>,</span>
<span>                                                <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Use of gghill_rarefaction requires a vector with named pcrs</span></span>
<span><span class="va">sampler</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">sampler</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hill_rarefaction.html">gghill_rarefaction</a></span><span class="op">(</span><span class="va">karst_euk.raref</span>, group<span class="op">=</span><span class="va">sampler</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"orange"</span>,<span class="st">"chartreuse"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"orange"</span>,<span class="st">"chartreuse"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"Sampler"</span><span class="op">)</span></span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/hill-1.png" width="672" style="display: block; margin: auto;"></p>
<p>At a first glance, it seems that Jean’s samples are more diverse in
abundant/intermediate sequences than Yvette’s ones.</p>
</div>
</div>
<div class="section level2">
<h2 id="flagging-spurious-motus">Flagging spurious MOTUs<a class="anchor" aria-label="anchor" href="#flagging-spurious-motus"></a>
</h2>
<div class="section level3">
<h3 id="potential-contaminants">Potential contaminants<a class="anchor" aria-label="anchor" href="#potential-contaminants"></a>
</h3>
<p>This dataset does not include negative controls to enable the
detection of potential contaminants objectively. Still, we identified
above some <em>pcrs</em> with very low diversity that may correspond to
<em>pcrs</em> reactions where addition of DNA template somehow failed,
and where reagents contaminants were amplified. Let’s see if these have
a particular composition when compared with the other samples. This can
be done by hijacking the <code>contaslayer</code> function as
follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set low-diversity pcrs as controls of e.g. pcrs</span></span>
<span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">control_type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">nb_motus</span><span class="op">&lt;</span><span class="fl">100</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"pcr"</span></span>
<span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">nb_motus</span><span class="op">&lt;</span><span class="fl">100</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"control"</span></span>
<span></span>
<span><span class="co"># run contaslayer</span></span>
<span><span class="va">karst_euk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/contaslayer.html">contaslayer</a></span><span class="op">(</span><span class="va">karst_euk</span>, </span>
<span>                         control_types <span class="op">=</span> <span class="st">"pcr"</span>,</span>
<span>                         output_col <span class="op">=</span> <span class="st">"not_a_pcr_conta"</span><span class="op">)</span></span></code></pre></div>
<p>And then display some basic information on these potential
contaminating sequences.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># print taxonomy of MOTUs most abundant in low-diversity samples</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">kableExtra</a></span><span class="op">)</span></span>
<span><span class="va">dt</span> <span class="op">&lt;-</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">[</span><span class="op">!</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_a_pcr_conta</span>,</span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"count"</span>,<span class="st">"best_identity"</span>,</span>
<span>                       <span class="st">"scientific_name"</span>, <span class="st">"sequence"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">dt</span><span class="op">$</span><span class="va">best_identity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">dt</span><span class="op">$</span><span class="va">best_identity</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dt</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"total # reads"</span>, <span class="st">"similarity to ref DB"</span>, <span class="st">"taxon name"</span>, <span class="st">"sequence"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">dt</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">dt</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,<span class="op">]</span>, row.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html" class="external-link">kable_styling</a></span><span class="op">(</span>bootstrap_options<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"striped"</span>, <span class="st">"hover"</span>, <span class="st">"condensed"</span><span class="op">)</span>, </span>
<span>                font_size <span class="op">=</span> <span class="fl">8</span>, full_width <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-striped table-hover table-condensed" style="font-size: 8px; width: auto !important; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
total # reads
</th>
<th style="text-align:right;">
similarity to ref DB
</th>
<th style="text-align:left;">
taxon name
</th>
<th style="text-align:left;">
sequence
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
laos_18Seuka_00044
</td>
<td style="text-align:right;">
2752
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Sordariomycetes
</td>
<td style="text-align:left;">
ctcaaacttccatccgcttgagcggatagtccctctaagaagccagcgtactgccaaagcaatacgggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00076
</td>
<td style="text-align:right;">
1768
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Liliopsida
</td>
<td style="text-align:left;">
ctcaaacttccgtggcctaaacggccatagtccctctaagaagctagctgcggagggatggctccgcatagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00048
</td>
<td style="text-align:right;">
738
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Pooideae
</td>
<td style="text-align:left;">
ctcaaacttccgtcgcctaaacggcgatagtccctctaagaagctagctgcggagggatggctccgcatagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00390
</td>
<td style="text-align:right;">
437
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Euclea crispa subsp. linearis
</td>
<td style="text-align:left;">
ctcaaacttccgtggcctgaaaggccatagtccctctaagaagctagctgcggagggtcgcctccgcatagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00379
</td>
<td style="text-align:right;">
400
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Brassicaceae
</td>
<td style="text-align:left;">
ctcaaacttccttggcctaaacggccatagtccctctaagaagccggccgtgaagggatgcctccacgtagctagttagcaggctgaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00098
</td>
<td style="text-align:right;">
266
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Euplotida
</td>
<td style="text-align:left;">
ctcaaacttccttgtggttgcacacaaagtccctctaagaagtacataccggaaaaccggctgactatttagcaggctaaggtctcgttcgttaa
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00366
</td>
<td style="text-align:right;">
248
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Basidiomycota
</td>
<td style="text-align:left;">
ctcaaacttccgtcagctaaacgctgacagtccctctaagaagccagcgaccagcaaaagccggccgggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00329
</td>
<td style="text-align:right;">
220
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Diplosoma gumavirens
</td>
<td style="text-align:left;">
ctcaaacttccattggctggaagccaatagtccctctaagaagccagccaccaaccatagtcgatggggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_00866
</td>
<td style="text-align:right;">
176
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Fungi
</td>
<td style="text-align:left;">
ctcaaacttccttcggcttgagccgaaagtccctctaagaagccagcgtactgccaaagcaatacgggctatttagcaggttaaggtctcgttcgttat
</td>
</tr>
<tr>
<td style="text-align:left;">
laos_18Seuka_01002
</td>
<td style="text-align:right;">
174
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
Vorticella campanula
</td>
<td style="text-align:left;">
ctcaaacttccatgtgattacatcacatagtccctctaagaagtgattcaaatttgaaataagaacactagttagcaggttaaggtctcgttcgttaa
</td>
</tr>
</tbody>
</table>
<p>From the taxonomy of these potential contaminant sequences, it is
difficult to say whether these sequences are true contaminants. One way
to further investigate this is to see how these sequences are
distributed across <em>pcrs</em>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">reads</span><span class="op">[</span>,<span class="op">!</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_a_pcr_conta</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tmp</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Var2</span>, y<span class="op">=</span><span class="va">Var1</span>, size<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">value</span><span class="op">==</span><span class="fl">0</span>, <span class="cn">NA</span>, <span class="va">value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>size <span class="op">=</span> <span class="st">"# reads"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/conta3-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The most abundant contaminant identified in the table above is
present in all samples and is, therefore, most likely a contaminant. In
addition, other potential contaminant sequences are present in 7-8
successive samples, which are likely in the same column in the PCR
plate. These features strongly suggest that these sequences are indeed
artefactual and should be excluded from the dataset.</p>
</div>
<div class="section level3">
<h3 id="flagging-spurious-motus-and-non-target-motus">Flagging spurious MOTUs and non-target MOTUs<a class="anchor" aria-label="anchor" href="#flagging-spurious-motus-and-non-target-motus"></a>
</h3>
<p>Non-target MOTUs cannot be detected in this particular dataset,
because the taxonomic assignments have been made using a reference
database that only contains eukaryotic sequences.</p>
<p>One can, however, still identify MOTUs whose sequence is too
dissimilar from references, by plotting the distribution of MOTU
similarity scores, weighted and unweighted by their relative
abundance.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the unweighted distribution of MOTUs similarity scores </span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">best_identity</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"grey"</span>, fill<span class="op">=</span><span class="st">"white"</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.8</span>, col<span class="op">=</span><span class="st">"orange"</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"% similarity against best match"</span>, y<span class="op">=</span><span class="st">"# MOTUs"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Same for the weighted distribution</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span>, </span>
<span>         <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">best_identity</span>, y <span class="op">=</span> <span class="va">..count..</span>, weight <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"grey"</span>, fill<span class="op">=</span><span class="st">"white"</span>, bins<span class="op">=</span><span class="fl">20</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.8</span>, col<span class="op">=</span><span class="st">"orange"</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"% similarity against best match"</span>, y<span class="op">=</span><span class="st">"# Reads"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine plots into one</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/ggdraw.html" class="external-link">ggdraw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html" class="external-link">draw_plot</a></span><span class="op">(</span><span class="va">a</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0</span>, width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html" class="external-link">draw_plot</a></span><span class="op">(</span><span class="va">b</span>, x<span class="op">=</span><span class="fl">0.5</span>, y<span class="op">=</span><span class="fl">0</span>, width <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/nontarget3-1.png" width="576" style="display: block; margin: auto;"></p>
<p>As in the <code>soil_euk</code> dataset, we may consider any MOTU as
degraded sequences if its sequence similarity is &lt; 80% against its
best match in the reference database.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Flag not degraded (TRUE) vs. potentially degraded sequences (FALSE)</span></span>
<span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_degraded</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">best_identity</span> <span class="op">&lt;</span> <span class="fl">0.8</span>, <span class="cn">F</span>, <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Proportion of each of these over total number of MOTUs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_degraded</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      FALSE       TRUE </span></span>
<span><span class="co">#&gt; 0.02210928 0.97789072</span></span>
<span></span>
<span><span class="co"># Intersection with other flags</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_a_pcr_conta</span>, </span>
<span>      <span class="va">karst_euk</span><span class="op">$</span><span class="va">motus</span><span class="op">$</span><span class="va">not_degraded</span><span class="op">)</span></span>
<span><span class="co">#&gt;        </span></span>
<span><span class="co">#&gt;         FALSE TRUE</span></span>
<span><span class="co">#&gt;   FALSE     0   40</span></span>
<span><span class="co">#&gt;   TRUE    174 7656</span></span></code></pre></div>
<p>In this case, no intersection is found between degraded MOTUs and
potential contaminants.</p>
</div>
</div>
<div class="section level2">
<h2 id="assessing-signal-reproducibility">Assessing signal reproducibility<a class="anchor" aria-label="anchor" href="#assessing-signal-reproducibility"></a>
</h2>
<p>Using the characteristics of the <code>karst_euk</code> dataset, one
can assess the signal reproducibility at different levels:<br>
- the DNA extract level<br>
- the composite sample level<br>
- the site level.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DNA extraction</span></span>
<span><span class="va">comp1</span> <span class="op">=</span> <span class="fu"><a href="../reference/pcrslayer.html">pcr_within_between</a></span><span class="op">(</span><span class="va">karst_euk</span>, replicates <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">extraction_id</span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_thresh</a></span><span class="op">(</span><span class="va">comp1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"between DNA extracts"</span>, <span class="st">"within DNA extracts"</span><span class="op">)</span>, </span>
<span>                       values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"coral2"</span>, <span class="st">"cyan3"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># composite sample</span></span>
<span><span class="va">comp2</span> <span class="op">=</span> <span class="fu"><a href="../reference/pcrslayer.html">pcr_within_between</a></span><span class="op">(</span><span class="va">karst_euk</span>, replicates <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_thresh</a></span><span class="op">(</span><span class="va">comp2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"between samples"</span>, <span class="st">"within DNA samples"</span><span class="op">)</span>, </span>
<span>                       values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"coral2"</span>, <span class="st">"cyan3"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># sampler</span></span>
<span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sampler_id</span> <span class="op">&lt;-</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">sampler</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sample_id</span>, </span>
<span>                                                             <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">comp3</span> <span class="op">=</span> <span class="fu"><a href="../reference/pcrslayer.html">pcr_within_between</a></span><span class="op">(</span><span class="va">karst_euk</span>, replicates <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sampler_id</span><span class="op">)</span></span>
<span></span>
<span><span class="va">comp3</span><span class="op">$</span><span class="va">bar_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">comp3</span><span class="op">$</span><span class="va">bar_dist</span>, <span class="fl">2</span><span class="op">)</span> <span class="co"># trick to plot a density from 1 value</span></span>
<span></span>
<span><span class="va">c</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_thresh</a></span><span class="op">(</span><span class="va">comp3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"between samplers"</span>, <span class="st">"within DNA samplers"</span><span class="op">)</span>, </span>
<span>                       values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"coral2"</span>, <span class="st">"cyan3"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/ggdraw.html" class="external-link">ggdraw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html" class="external-link">draw_plot</a></span><span class="op">(</span><span class="va">a</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0.6</span>, height <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html" class="external-link">draw_plot</a></span><span class="op">(</span><span class="va">b</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0.3</span>,  height <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_plot.html" class="external-link">draw_plot</a></span><span class="op">(</span><span class="va">c</span>, x<span class="op">=</span><span class="fl">0</span>, y<span class="op">=</span><span class="fl">0</span>,  height <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/pcrsalyer1-1.png" width="480" style="display: block; margin: auto;"></p>
<p>These plots suggest that the PCR and extraction reproducibility is
high. By contrast, reproducibility per sampler is low, which suggests
that the biological differences between samples are stronger than
potential effects introduced by the sampler identity.</p>
<p>Let’s verify this with an ordination visualisation by hijacking the
<code>check_pcr_repl</code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Distinguish between pcrs obtained from the different sample sites</span></span>
<span><span class="va">mds</span> <span class="op">=</span> <span class="fu"><a href="../reference/pcrslayer.html">check_pcr_repl</a></span><span class="op">(</span><span class="va">karst_euk</span>, </span>
<span>                     <span class="co"># display sites with different colors</span></span>
<span>                     groups <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Site</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sample_id</span>,</span>
<span>                                                           <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">karst_euk</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                     <span class="co"># use funcpcr to differentiate samplers instead of good vs. failed pcrs.</span></span>
<span>                     funcpcr <span class="op">=</span> <span class="va">karst_euk</span><span class="op">$</span><span class="va">pcrs</span><span class="op">$</span><span class="va">sampler_id</span> <span class="op">==</span> <span class="st">"Yvette"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">mds</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Sites"</span>, shape<span class="op">=</span><span class="st">"Sampler"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_shape.html" class="external-link">scale_shape</a></span><span class="op">(</span>labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Jean"</span>, <span class="st">"Yvette"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="metabaRF-vignette_karst_files/figure-html/pcrsalyer3-1.png" width="480" style="display: block; margin: auto;"></p>
<p>The variability across sites is hence far larger than between
samplers and technical replicates. But the sampler effect is not
negligible. One can observe, however, some PCR outliers (site D1-2
sampled by Jean, or site C sampled by Yvette) that can be flagged with
the function <code>pcrslayer</code>. We invite readers to refer to the
<a href="https://metabarfactory.github.io/metabaR/articles/metabaRF-vignette.html">“Let’s
metabaR” tutorial</a> for further advices on the use of this
function.</p>
</div>
<div class="section level2">
<h2 id="data-final-cleaning-and-aggregation">Data final cleaning and aggregation<a class="anchor" aria-label="anchor" href="#data-final-cleaning-and-aggregation"></a>
</h2>
<p>The above represents a succinct description of the data cleaning
process for the <code>karst_euk</code> dataset. Note that not all steps
are shown here, and we refer the user to the <a href="https://metabarfactory.github.io/metabaR/articles/metabaRF-vignette.html">“Let’s
metabaR” tutorial</a>, based on the <code>soil_euk</code> dataset for a
complete process of signal flagging and cleaning, including the last
steps of the process, i.e. removal of flagged MOTUs/pcrs and data
aggregation.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-boyer:2016:00" class="csl-entry">
Boyer, F., C. Mercier, A. Bonin, Y. Le Bras, P. Taberlet, and E.
Coissac. 2016. <span>“OBITools : A Unix- Inspired Software Package for
DNA Metabarcoding.”</span> <em>Molecular Ecology Resources</em> 16 (1):
176–82.
</div>
<div id="ref-taberlet2018environmental" class="csl-entry">
Taberlet, Pierre, Aurélie Bonin, Lucie Zinger, and Eric Coissac. 2018.
<em>Environmental DNA: For Biodiversity Research and Monitoring</em>.
Oxford University Press.
</div>
<div id="ref-zinger2019body" class="csl-entry">
Zinger, Lucie, Pierre Taberlet, Heidy Schimann, Aurélie Bonin, Frédéric
Boyer, Marta De Barba, Philippe Gaucher, et al. 2019. <span>“Body Size
Determines Soil Community Assembly in a Tropical Forest.”</span>
<em>Molecular Ecology</em> 28 (3): 528–43.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Lucie Zinger, Frédéric Boyer, Clément Lionnet, Anne-Sophie Benoiston.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
